// @progress/kendo-angular-buttons/esm2022/floatingactionbutton/floatingactionbutton.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_FLOATINGACTIONBUTTON } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./floatingactionbutton.component";
import * as i2 from "./templates/dial-item-template.directive";
import * as i3 from "./templates/fab-template.directive";
/**
 * Represents the NgModule definition for the FloatingActionButton component.
 *
 * Required for adding FloatingActionButton features in NgModule-based Angular applications.
 *
 * @example
 * ```ts
 * import { FloatingActionButtonModule } from '@progress/kendo-angular-buttons';
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 * import { BrowserModule } from '@angular/platform-browser';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports: [BrowserModule, FloatingActionButtonModule],
 *     bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class FloatingActionButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: FloatingActionButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: FloatingActionButtonModule, imports: [i1.FloatingActionButtonComponent, i2.DialItemTemplateDirective, i3.FloatingActionButtonTemplateDirective], exports: [i1.FloatingActionButtonComponent, i2.DialItemTemplateDirective, i3.FloatingActionButtonTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: FloatingActionButtonModule, providers: [IconsService, PopupService, ResizeBatchService], imports: [i1.FloatingActionButtonComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: FloatingActionButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_FLOATINGACTIONBUTTON],
                    imports: [...KENDO_FLOATINGACTIONBUTTON],
                    providers: [IconsService, PopupService, ResizeBatchService]
                }]
        }] });
