// @progress/kendo-angular-treeview/esm2022/treeview.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_TREEVIEW } from './directives';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { DialogContainerService, DialogService, WindowContainerService, WindowService } from '@progress/kendo-angular-dialog';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import * as i0 from "@angular/core";
import * as i1 from "./treeview.component";
import * as i2 from "./node-template.directive";
import * as i3 from "./check.directive";
import * as i4 from "./disable.directive";
import * as i5 from "./expand.directive";
import * as i6 from "./selection/select.directive";
import * as i7 from "./hierarchy-binding.directive";
import * as i8 from "./flat-binding.directive";
import * as i9 from "./drag-and-drop/drag-and-drop.directive";
import * as i10 from "./drag-and-drop/drag-clue/drag-clue-template.directive";
import * as i11 from "./drag-and-drop/drop-hint/drop-hint-template.directive";
import * as i12 from "./drag-and-drop/drag-and-drop-editing.directive";
import * as i13 from "./load-more/load-more.directive";
import * as i14 from "./load-more/load-more-button-template.directive";
import * as i15 from "./localization/custom-messages.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi']) definition for the TreeView component.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { TreeViewModule } from '@progress/kendo-angular-treeview';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   imports: [BrowserModule, TreeViewModule],
 *   declarations: [AppComponent],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class TreeViewModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TreeViewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TreeViewModule, imports: [i1.TreeViewComponent, i2.NodeTemplateDirective, i3.CheckDirective, i4.DisableDirective, i5.ExpandDirective, i6.SelectDirective, i7.HierarchyBindingDirective, i8.FlatDataBindingDirective, i9.DragAndDropDirective, i10.DragClueTemplateDirective, i11.DropHintTemplateDirective, i12.DragAndDropEditingDirective, i13.LoadMoreDirective, i14.LoadMoreButtonTemplateDirective, i15.CustomMessagesComponent], exports: [i1.TreeViewComponent, i2.NodeTemplateDirective, i3.CheckDirective, i4.DisableDirective, i5.ExpandDirective, i6.SelectDirective, i7.HierarchyBindingDirective, i8.FlatDataBindingDirective, i9.DragAndDropDirective, i10.DragClueTemplateDirective, i11.DropHintTemplateDirective, i12.DragAndDropEditingDirective, i13.LoadMoreDirective, i14.LoadMoreButtonTemplateDirective, i15.CustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TreeViewModule, providers: [
            IconsService,
            PopupService,
            ResizeBatchService,
            DialogContainerService,
            DialogService,
            WindowService,
            WindowContainerService
        ], imports: [i1.TreeViewComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TreeViewModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_TREEVIEW],
                    exports: [...KENDO_TREEVIEW],
                    providers: [
                        IconsService,
                        PopupService,
                        ResizeBatchService,
                        DialogContainerService,
                        DialogService,
                        WindowService,
                        WindowContainerService
                    ]
                }]
        }] });
