// @progress/kendo-angular-treeview/esm2022/drag-and-drop/models/drop-action.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/**
 * Represents the attempted drop action during dragging.
 * Passed as `action` value to the [`kendoTreeViewDragClueTemplate`](slug:api_treeview_dragcluetemplatedirective) directive.
 * By default, this value defines the rendered icon in the drag clue.
 */
export var DropAction;
(function (DropAction) {
    /**
     * The dragged item is added as a child to the target node.
     */
    DropAction[DropAction["Add"] = 0] = "Add";
    /**
     * The dragged item is inserted above the target node at the same level.
     */
    DropAction[DropAction["InsertTop"] = 1] = "InsertTop";
    /**
     * The dragged item is inserted below the target node at the same level.
     */
    DropAction[DropAction["InsertBottom"] = 2] = "InsertBottom";
    /**
     * The dragged item is inserted in the middle position relative to the target node.
     */
    DropAction[DropAction["InsertMiddle"] = 3] = "InsertMiddle";
    /**
     * The drop action is not valid for the current target location.
     */
    DropAction[DropAction["Invalid"] = 4] = "Invalid";
})(DropAction || (DropAction = {}));
