// @progress/kendo-angular-charts/esm2022/stock-chart/navigator/category-axis/notes.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ConfigurationService } from '../../../common/configuration.service';
import { CategoryAxisNotesComponent } from '../../../chart/category-axis-item/notes.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../common/configuration.service";
/**
 * Represents notes configuration of the navigator category axis.
 *
 * @example
 * ```html
 * <kendo-stockchart>
 *  <kendo-chart-navigator>
 *   <kendo-chart-navigator-category-axis>
 *    <kendo-chart-navigator-category-axis-notes [visible]="true">
 *    </kendo-chart-navigator-category-axis-notes>
 *   </kendo-chart-navigator-category-axis>
 *  </kendo-chart-navigator>
 * </kendo-stockchart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link NavigatorCategoryAxisNotesIconComponent} and {@link NavigatorCategoryAxisNotesLabelComponent}.
 */
export class NavigatorCategoryAxisNotesComponent extends CategoryAxisNotesComponent {
    configurationService;
    constructor(configurationService) {
        super(configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorCategoryAxisNotesComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: NavigatorCategoryAxisNotesComponent, isStandalone: true, selector: "kendo-chart-navigator-category-axis-notes", usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorCategoryAxisNotesComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-navigator-category-axis-notes',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; } });
