// @progress/kendo-angular-charts/esm2022/stock-chart/navigator/category-axis/select.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ConfigurationService } from '../../../common/configuration.service';
import { CategoryAxisSelectComponent } from '../../../chart/category-axis-item/select.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../common/configuration.service";
/**
 * Specifies the selected axis range.
 * When set, axis selection is enabled. The range uses zero-based indexes.
 * Selects categories with indexes in the range (`select.from`, `select.to`).
 * The last category in the range is not included in the selection.
 * If categories are dates, specify the range with date values.
 *
 * @example
 * ```typescript
 * <kendo-chart-navigator-category-axis-select [from]="1" [to]="3"></kendo-chart-navigator-category-axis-select>
 * ```
 */
export class NavigatorCategoryAxisSelectComponent extends CategoryAxisSelectComponent {
    configurationService;
    constructor(configurationService) {
        super(configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorCategoryAxisSelectComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: NavigatorCategoryAxisSelectComponent, isStandalone: true, selector: "kendo-chart-navigator-category-axis-select", usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorCategoryAxisSelectComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-navigator-category-axis-select',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; } });
