// @progress/kendo-angular-treeview/esm2022/load-more/load-more-button-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Optional, Directive, TemplateRef } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * Represents a directive for customizing the load more button in the TreeView.
 *
 * To define the template, nest an `<ng-template>` tag with the `kendoTreeViewLoadMoreButtonTemplate` directive inside a `<kendo-treeview>` tag
 * ([see example](slug:loadmorebutton_treeview#button-template)).
 *
 * The template context provides the following variable:
 * - `let-index="index"`&mdash;The hierarchical index of the load more button node.
 *
 * @example
 * ```html
 * <kendo-treeview>
 *   <ng-template kendoTreeViewLoadMoreButtonTemplate let-index="index">
 *     Load more at {{ index }}
 *   </ng-template>
 * </kendo-treeview>
 * ```
 */
export class LoadMoreButtonTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LoadMoreButtonTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: LoadMoreButtonTemplateDirective, isStandalone: true, selector: "[kendoTreeViewLoadMoreButtonTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LoadMoreButtonTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoTreeViewLoadMoreButtonTemplate]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }]; } });
