// @progress/kendo-angular-dropdowns/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { PrefixTemplateDirective, SuffixTemplateDirective, SeparatorComponent, ToggleButtonTabStopDirective } from "@progress/kendo-angular-common";
import { AutoCompleteComponent } from "./autocomplete/autocomplete.component";
import { FilterDirective } from "./common/filtering/filter.directive";
import { CustomMessagesComponent } from "./common/localization/custom-messages.component";
import { FixedGroupTemplateDirective } from "./common/templates/fixed-group-template.directive";
import { FooterTemplateDirective } from "./common/templates/footer-template.directive";
import { GroupTemplateDirective } from "./common/templates/group-template.directive";
import { HeaderTemplateDirective } from "./common/templates/header-template.directive";
import { ItemTemplateDirective } from "./common/templates/item-template.directive";
import { NoDataTemplateDirective } from "./common/templates/no-data-template.directive";
import { ComboBoxComponent } from "./comboboxes/combobox.component";
import { MultiColumnComboBoxComponent } from "./comboboxes/multicolumncombobox.component";
import { ComboBoxColumnComponent } from "./comboboxes/combobox-column/combobox-column.component";
import { ColumnHeaderTemplateDirective } from "./comboboxes/combobox-column/column-header-template.directive";
import { ColumnCellTemplateDirective } from "./comboboxes/combobox-column/column-cell-template.directive";
import { DropDownListComponent } from "./dropdownlist/dropdownlist.component";
import { ValueTemplateDirective } from "./common/templates/value-template.directive";
import { MultiSelectComponent } from "./multiselect/multiselect.component";
import { SummaryTagDirective } from "./multiselect/summary-tag.directive";
import { CustomItemTemplateDirective } from "./common/templates/custom-item-template.directive";
import { GroupTagTemplateDirective } from "./common/templates/group-tag-template.directive";
import { TagTemplateDirective } from "./common/templates/tag-template.directive";
import { DropDownTreeComponent } from "./dropdowntrees/dropdowntree.component";
import { MultiSelectTreeComponent } from "./dropdowntrees/multiselecttree.component";
import { DropDownTreeFlatBindingDirective } from "./dropdowntrees/data-binding/dropdowntree/flat-binding.directive";
import { DropDownTreeHierarchyBindingDirective } from "./dropdowntrees/data-binding/dropdowntree/hierarchy-binding.directive";
import { MultiSelectTreeFlatBindingDirective } from "./dropdowntrees/data-binding/multiselecttree/flat-binding.directive";
import { MultiSelectTreeHierarchyBindingDirective } from "./dropdowntrees/data-binding/multiselecttree/hierarchy-binding.directive";
import { DropDownTreesExpandDirective } from "./dropdowntrees/expanded-state/expand.directive";
import { NodeTemplateDirective } from "./dropdowntrees/templates/node-template.directive";
import { MultiSelectTreeSummaryTagDirective } from "./dropdowntrees/summary-tag/summary-tag.directive";
/**
 * Utility array that contains all `AutoComplete` related components and directives
 */
export const KENDO_AUTOCOMPLETE = [
    AutoCompleteComponent,
    FixedGroupTemplateDirective,
    FooterTemplateDirective,
    GroupTemplateDirective,
    HeaderTemplateDirective,
    ItemTemplateDirective,
    NoDataTemplateDirective,
    FilterDirective,
    CustomMessagesComponent,
    SuffixTemplateDirective,
    PrefixTemplateDirective,
    SeparatorComponent
];
/**
 * Utility array that contains all `ComboBox` related components and directives
 */
export const KENDO_COMBOBOX = [
    ComboBoxComponent,
    FixedGroupTemplateDirective,
    FooterTemplateDirective,
    GroupTemplateDirective,
    HeaderTemplateDirective,
    ItemTemplateDirective,
    NoDataTemplateDirective,
    FilterDirective,
    CustomMessagesComponent,
    SuffixTemplateDirective,
    PrefixTemplateDirective,
    SeparatorComponent,
    ToggleButtonTabStopDirective
];
/**
 * Utility array that contains all `MultiColumnComboBox` related components and directives
 */
export const KENDO_MULTICOLUMNCOMBOBOX = [
    MultiColumnComboBoxComponent,
    ComboBoxColumnComponent,
    ColumnHeaderTemplateDirective,
    ColumnCellTemplateDirective,
    FixedGroupTemplateDirective,
    FooterTemplateDirective,
    GroupTemplateDirective,
    HeaderTemplateDirective,
    ItemTemplateDirective,
    NoDataTemplateDirective,
    FilterDirective,
    CustomMessagesComponent,
    SuffixTemplateDirective,
    PrefixTemplateDirective,
    SeparatorComponent,
    ToggleButtonTabStopDirective
];
/**
 * Utility array that contains all `DropDownList` related components and directives
 */
export const KENDO_DROPDOWNLIST = [
    DropDownListComponent,
    FixedGroupTemplateDirective,
    FooterTemplateDirective,
    GroupTemplateDirective,
    HeaderTemplateDirective,
    ItemTemplateDirective,
    ValueTemplateDirective,
    NoDataTemplateDirective,
    FilterDirective,
    CustomMessagesComponent
];
/**
 * Utility array that contains all `MultiSelect` related components and directives
 */
export const KENDO_MULTISELECT = [
    MultiSelectComponent,
    SummaryTagDirective,
    CustomItemTemplateDirective,
    FixedGroupTemplateDirective,
    FooterTemplateDirective,
    GroupTemplateDirective,
    GroupTagTemplateDirective,
    TagTemplateDirective,
    HeaderTemplateDirective,
    ItemTemplateDirective,
    NoDataTemplateDirective,
    FilterDirective,
    CustomMessagesComponent,
    SuffixTemplateDirective,
    PrefixTemplateDirective,
    SeparatorComponent
];
/**
 * Utility array that contains all `DropDownTree` related components and directives
 */
export const KENDO_DROPDOWNTREE = [
    DropDownTreeComponent,
    DropDownTreeFlatBindingDirective,
    DropDownTreeHierarchyBindingDirective,
    DropDownTreesExpandDirective,
    NodeTemplateDirective,
    HeaderTemplateDirective,
    FooterTemplateDirective,
    ValueTemplateDirective,
    NoDataTemplateDirective,
    CustomMessagesComponent
];
/**
 * Utility array that contains all `MultiSelectTree` related components and directives
 */
export const KENDO_MULTISELECTTREE = [
    MultiSelectTreeComponent,
    MultiSelectTreeFlatBindingDirective,
    MultiSelectTreeHierarchyBindingDirective,
    DropDownTreesExpandDirective,
    NodeTemplateDirective,
    MultiSelectTreeSummaryTagDirective,
    GroupTagTemplateDirective,
    TagTemplateDirective,
    HeaderTemplateDirective,
    FooterTemplateDirective,
    ValueTemplateDirective,
    NoDataTemplateDirective,
    CustomMessagesComponent
];
/**
 * Utility array that contains all `@progress/kendo-angular-dropdowns` related components and directives
 */
export const KENDO_DROPDOWNS = [
    ...KENDO_AUTOCOMPLETE,
    ...KENDO_COMBOBOX,
    ...KENDO_MULTICOLUMNCOMBOBOX,
    ...KENDO_DROPDOWNLIST,
    ...KENDO_MULTISELECT,
    ...KENDO_DROPDOWNTREE,
    ...KENDO_MULTISELECTTREE
];
