// @progress/kendo-angular-charts/esm2022/chart/pane/title.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the title configuration for the Chart panes.
 */
export class PanesTitleComponent extends SettingsComponent {
    configurationService;
    /**
     * Sets the background color of the title.
     */
    background;
    /**
     * Specifies the border configuration of the title.
     */
    border;
    /**
     * Sets the text color of the title.
     */
    color;
    /**
     * Defines the font style of the title.
     * @default '700 18px sans-serif'
     */
    font;
    /**
     * Specifies the margin of the title. A numeric value sets all margins.
     * @default 5
     */
    margin;
    /**
     * Specifies the position of the title.
     * @default 'center'
     */
    position;
    /**
     * Sets the text content of the title.
     */
    text;
    /**
     * Determines whether the Chart displays the pane title.
     * @default true
     */
    visible;
    /**
     * Provides a function that creates a custom visual for the title.
     */
    visual;
    constructor(configurationService) {
        super('title', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PanesTitleComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: PanesTitleComponent, isStandalone: true, selector: "kendo-chart-pane-title", inputs: { background: "background", border: "border", color: "color", font: "font", margin: "margin", position: "position", text: "text", visible: "visible", visual: "visual" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PanesTitleComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-pane-title',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], font: [{
                type: Input
            }], margin: [{
                type: Input
            }], position: [{
                type: Input
            }], text: [{
                type: Input
            }], visible: [{
                type: Input
            }], visual: [{
                type: Input
            }] } });
