// @progress/kendo-angular-utils/esm2022/drag-and-drop.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DRAGANDDROP } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./drag-and-drop/dragtarget.directive";
import * as i2 from "./drag-and-drop/draghandle.directive";
import * as i3 from "./drag-and-drop/droptarget.directive";
import * as i4 from "./drag-and-drop/dragtarget-container.directive";
import * as i5 from "./drag-and-drop/droptarget-container.directive";
import * as i6 from "./drag-and-drop/hint.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmodules'])
 * definition for the Drag and Drop directives.
 *
 * @example
 * ```ts
 * import { DragAndDropModule } from '@progress/kendo-angular-utils';
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, DragAndDropModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class DragAndDropModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DragAndDropModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DragAndDropModule, imports: [i1.DragTargetDirective, i2.DragHandleDirective, i3.DropTargetDirective, i4.DragTargetContainerDirective, i5.DropTargetContainerDirective, i6.HintComponent], exports: [i1.DragTargetDirective, i2.DragHandleDirective, i3.DropTargetDirective, i4.DragTargetContainerDirective, i5.DropTargetContainerDirective, i6.HintComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DragAndDropModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DragAndDropModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DRAGANDDROP],
                    imports: [...KENDO_DRAGANDDROP]
                }]
        }] });
