// @progress/kendo-angular-charts/esm2022/chart/value-axis-item/crosshair.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the crosshair configuration options of the Chart value axis ([see example]({% slug crosshairs_chart_charts %})).
 *
 * @example
 * ```html
 * <kendo-chart>
 *  <kendo-chart-value-axis>
 *   <kendo-chart-value-axis-item>
 *    <kendo-chart-value-axis-item-crosshair [visible]="true" color="#ff0000" [width]="2">
 *     <kendo-chart-value-axis-item-crosshair-tooltip format="c"></kendo-chart-value-axis-item-crosshair-tooltip>
 *    </kendo-chart-value-axis-item-crosshair>
 *   </kendo-chart-value-axis-item>
 *  </kendo-chart-value-axis>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link ValueAxisCrosshairTooltipComponent}.
 */
export class ValueAxisCrosshairComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the color of the crosshair.
     */
    color;
    /**
     * Specifies the opacity of the crosshair.
     *
     * @default 1
     */
    opacity;
    /**
     * Determines whether the Chart displays the value axis crosshair.
     *
     * @default false
     */
    visible;
    /**
     * Specifies the width of the crosshair in pixels.
     *
     * @default 1
     */
    width;
    // These options are also available as child components
    /**
     * Specifies the tooltip configuration of the crosshair.
     */
    tooltip;
    constructor(configurationService) {
        super('crosshair', configurationService);
        this.configurationService = configurationService;
        this.markAsVisible();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ValueAxisCrosshairComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: ValueAxisCrosshairComponent, isStandalone: true, selector: "kendo-chart-value-axis-item-crosshair", inputs: { color: "color", opacity: "opacity", visible: "visible", width: "width", tooltip: "tooltip" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ValueAxisCrosshairComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-value-axis-item-crosshair',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { color: [{
                type: Input
            }], opacity: [{
                type: Input
            }], visible: [{
                type: Input
            }], width: [{
                type: Input
            }], tooltip: [{
                type: Input
            }] } });
