// @progress/kendo-angular-dropdowns/esm2022/multiselect/multiselect.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_MULTISELECT } from '../directives';
import { KENDO_ADORNMENTS, ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { AdaptiveService } from '@progress/kendo-angular-utils';
import * as i0 from "@angular/core";
import * as i1 from "./multiselect.component";
import * as i2 from "./summary-tag.directive";
import * as i3 from "../common/templates/custom-item-template.directive";
import * as i4 from "../common/templates/fixed-group-template.directive";
import * as i5 from "../common/templates/footer-template.directive";
import * as i6 from "../common/templates/group-template.directive";
import * as i7 from "../common/templates/group-tag-template.directive";
import * as i8 from "../common/templates/tag-template.directive";
import * as i9 from "../common/templates/header-template.directive";
import * as i10 from "../common/templates/item-template.directive";
import * as i11 from "../common/templates/no-data-template.directive";
import * as i12 from "../common/filtering/filter.directive";
import * as i13 from "../common/localization/custom-messages.component";
import * as i14 from "@progress/kendo-angular-common";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Required for adding all MultiSelect features in NgModule-based Angular applications.
 *
 * This module exports all MultiSelect-related components and directives:
 * - `MultiSelectComponent`&mdash;The main MultiSelect component class.
 * - `SummaryTagDirective`&mdash;The summary tag directive for MultiSelect.
 * - `ItemTemplateDirective`&mdash;The item template directive.
 * - `CustomItemTemplateDirective`&mdash;The custom item template directive.
 * - `TagTemplateDirective`&mdash;The tag template directive.
 * - `GroupTagTemplateDirective`&mdash;The group tag template directive.
 * - `SummaryTagTemplateDirective`&mdash;The summary tag template directive.
 * - `HeaderTemplateDirective`&mdash;The header template directive.
 * - `FooterTemplateDirective`&mdash;The footer template directive.
 * - `NoDataTemplateDirective`&mdash;The no-data template directive.
 * - `FilterDirective`&mdash;The filter directive.
 * - `CustomMessagesComponent`&mdash;The custom messages component.
 * - `GroupTemplateDirective`&mdash;The group template directive.
 * - `FixedGroupTemplateDirective`&mdash;The fixed group template directive.
 * - `SuffixTemplateDirective`&mdash;The suffix template directive.
 * - `PrefixTemplateDirective`&mdash;The prefix template directive.
 *
 * ```ts
 * import { MultiSelectModule } from '@progress/kendo-angular-dropdowns';
 * @NgModule({
 *  imports: [ MultiSelectModule ],
 *  declarations: [ AppComponent ],
 *  bootstrap: [ AppComponent ],
 * })
 * ```
 */
export class MultiSelectModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MultiSelectModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: MultiSelectModule, imports: [i1.MultiSelectComponent, i2.SummaryTagDirective, i3.CustomItemTemplateDirective, i4.FixedGroupTemplateDirective, i5.FooterTemplateDirective, i6.GroupTemplateDirective, i7.GroupTagTemplateDirective, i8.TagTemplateDirective, i9.HeaderTemplateDirective, i10.ItemTemplateDirective, i11.NoDataTemplateDirective, i12.FilterDirective, i13.CustomMessagesComponent, i14.SuffixTemplateDirective, i14.PrefixTemplateDirective, i14.SeparatorComponent, i14.PrefixTemplateDirective, i14.SuffixTemplateDirective, i14.SeparatorComponent], exports: [i1.MultiSelectComponent, i2.SummaryTagDirective, i3.CustomItemTemplateDirective, i4.FixedGroupTemplateDirective, i5.FooterTemplateDirective, i6.GroupTemplateDirective, i7.GroupTagTemplateDirective, i8.TagTemplateDirective, i9.HeaderTemplateDirective, i10.ItemTemplateDirective, i11.NoDataTemplateDirective, i12.FilterDirective, i13.CustomMessagesComponent, i14.SuffixTemplateDirective, i14.PrefixTemplateDirective, i14.SeparatorComponent, i14.PrefixTemplateDirective, i14.SuffixTemplateDirective, i14.SeparatorComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MultiSelectModule, providers: [
            PopupService,
            ResizeBatchService,
            IconsService,
            AdaptiveService
        ], imports: [i1.MultiSelectComponent, i14.SeparatorComponent, i14.SeparatorComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: MultiSelectModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_MULTISELECT, ...KENDO_ADORNMENTS],
                    exports: [...KENDO_MULTISELECT, ...KENDO_ADORNMENTS],
                    providers: [
                        PopupService,
                        ResizeBatchService,
                        IconsService,
                        AdaptiveService
                    ]
                }]
        }] });
