// @progress/kendo-angular-navigation/esm2022/breadcrumb.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { KENDO_BREADCRUMB } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./breadcrumb/breadcrumb.component";
import * as i2 from "./breadcrumb/template-directives/item-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the BreadCrumb component.
 *
 * @example
 * ```typescript
 * import { BreadCrumbModule } from '@progress/kendo-angular-navigation';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, BreadCrumbModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class BreadCrumbModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: BreadCrumbModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: BreadCrumbModule, imports: [i1.BreadCrumbComponent, i2.BreadCrumbItemTemplateDirective], exports: [i1.BreadCrumbComponent, i2.BreadCrumbItemTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: BreadCrumbModule, providers: [IconsService, ResizeBatchService], imports: [i1.BreadCrumbComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: BreadCrumbModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_BREADCRUMB],
                    imports: [...KENDO_BREADCRUMB],
                    providers: [IconsService, ResizeBatchService]
                }]
        }] });
