// @progress/kendo-angular-navigation/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ActionSheetViewComponent } from "./actionsheet/actionsheet-view.component";
import { ActionSheetComponent } from "./actionsheet/actionsheet.component";
import { ActionSheetContentTemplateDirective, ActionSheetFooterTemplateDirective, ActionSheetHeaderTemplateDirective, ActionSheetItemTemplateDirective, ActionSheetTemplateDirective } from "./actionsheet/models";
import { AppBarSectionComponent } from "./appbar/appbar-section.component";
import { AppBarSpacerComponent } from "./appbar/appbar-spacer.component";
import { AppBarComponent } from "./appbar/appbar.component";
import { BottomNavigationComponent } from "./bottomnavigation/bottomnavigation.component";
import { BottomNavigationItemTemplateDirective } from "./bottomnavigation/templates/item-template.directive";
import { BreadCrumbComponent } from "./breadcrumb/breadcrumb.component";
import { BreadCrumbItemTemplateDirective } from "./breadcrumb/template-directives/item-template.directive";
/**
* Use this utility array to access all ActionSheet-related components and directives in a standalone Angular component.
*
* @example
* ```typescript
* import { Component } from '@angular/core';
* import { KENDO_ACTIONSHEET } from '@progress/kendo-angular-navigation';
*
* @Component({
*   selector: 'my-app',
*   standalone: true,
*   imports: [KENDO_ACTIONSHEET],
*   template: `
*     <kendo-actionsheet [items]="items">
*     </kendo-actionsheet>
*   `
* })
* export class AppComponent {}
* ```
*/
export const KENDO_ACTIONSHEET = [
    ActionSheetComponent,
    ActionSheetViewComponent,
    ActionSheetHeaderTemplateDirective,
    ActionSheetItemTemplateDirective,
    ActionSheetContentTemplateDirective,
    ActionSheetFooterTemplateDirective,
    ActionSheetTemplateDirective
];
/**
* Use this utility array to access all AppBar-related components and directives in a standalone Angular component.
*
* @example
* ```typescript
* import { Component } from '@angular/core';
* import { KENDO_APPBAR } from '@progress/kendo-angular-navigation';
*
* @Component({
*   selector: 'my-app',
*   standalone: true,
*   imports: [KENDO_APPBAR],
*   template: `
*     <kendo-appbar>
*       <!-- AppBar content -->
*     </kendo-appbar>
*   `
* })
* export class AppComponent {}
* ```
*/
export const KENDO_APPBAR = [
    AppBarComponent,
    AppBarSectionComponent,
    AppBarSpacerComponent
];
/**
* Use this utility array to access all BottomNavigation-related components and directives in a standalone Angular component.
*
* @example
* ```typescript
* import { Component } from '@angular/core';
* import { KENDO_BOTTOMNAVIGATION } from '@progress/kendo-angular-navigation';
*
* @Component({
*   selector: 'my-app',
*   standalone: true,
*   imports: [KENDO_BOTTOMNAVIGATION],
*   template: `
*     <kendo-bottomnavigation [items]="items">
*     </kendo-bottomnavigation>
*   `
* })
* export class AppComponent {}
* ```
*/
export const KENDO_BOTTOMNAVIGATION = [
    BottomNavigationComponent,
    BottomNavigationItemTemplateDirective
];
/**
* Use this utility array to access all BreadCrumb-related components and directives in a standalone Angular component.
*
* @example
* ```typescript
* import { Component } from '@angular/core';
* import { KENDO_BREADCRUMB } from '@progress/kendo-angular-navigation';
*
* @Component({
*   selector: 'my-app',
*   standalone: true,
*   imports: [KENDO_BREADCRUMB],
*   template: `
*     <kendo-breadcrumb [items]="items">
*     </kendo-breadcrumb>
*   `
* })
* export class AppComponent {}
* ```
*/
export const KENDO_BREADCRUMB = [
    BreadCrumbComponent,
    BreadCrumbItemTemplateDirective
];
/**
* Use this utility array to access all `@progress/kendo-angular-navigation`-related components and directives in a standalone Angular component.
*
* @example
* ```typescript
* import { Component } from '@angular/core';
* import { KENDO_NAVIGATION } from '@progress/kendo-angular-navigation';
*
* @Component({
*   selector: 'my-app',
*   standalone: true,
*   imports: [KENDO_NAVIGATION],
*   template: `
*     <kendo-appbar>
*        <!-- AppBar content -->
*     </kendo-appbar>
*     <kendo-breadcrumb [items]="breadcrumbItems">
*     </kendo-breadcrumb>
*   `
* })
* export class AppComponent {}
* ```
*/
export const KENDO_NAVIGATION = [
    ...KENDO_ACTIONSHEET,
    ...KENDO_APPBAR,
    ...KENDO_BOTTOMNAVIGATION,
    ...KENDO_BREADCRUMB
];
