// @progress/kendo-angular-navigation/esm2022/navigation.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { KENDO_NAVIGATION } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./actionsheet/actionsheet.component";
import * as i2 from "./actionsheet/actionsheet-view.component";
import * as i3 from "./actionsheet/templates/header-template.directive";
import * as i4 from "./actionsheet/templates/item-template.directive";
import * as i5 from "./actionsheet/templates/content-template.directive";
import * as i6 from "./actionsheet/templates/footer-template.directive";
import * as i7 from "./actionsheet/templates/actionsheet-template";
import * as i8 from "./appbar/appbar.component";
import * as i9 from "./appbar/appbar-section.component";
import * as i10 from "./appbar/appbar-spacer.component";
import * as i11 from "./bottomnavigation/bottomnavigation.component";
import * as i12 from "./bottomnavigation/templates/item-template.directive";
import * as i13 from "./breadcrumb/breadcrumb.component";
import * as i14 from "./breadcrumb/template-directives/item-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [`NgModule`](link:site.data.urls.angular['ngmoduleapi'])
 * definition for the Navigation components.
 *
 * @example
 * ```typescript
 * import { NavigationModule } from '@progress/kendo-angular-navigation';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, NavigationModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class NavigationModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigationModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: NavigationModule, imports: [i1.ActionSheetComponent, i2.ActionSheetViewComponent, i3.ActionSheetHeaderTemplateDirective, i4.ActionSheetItemTemplateDirective, i5.ActionSheetContentTemplateDirective, i6.ActionSheetFooterTemplateDirective, i7.ActionSheetTemplateDirective, i8.AppBarComponent, i9.AppBarSectionComponent, i10.AppBarSpacerComponent, i11.BottomNavigationComponent, i12.BottomNavigationItemTemplateDirective, i13.BreadCrumbComponent, i14.BreadCrumbItemTemplateDirective], exports: [i1.ActionSheetComponent, i2.ActionSheetViewComponent, i3.ActionSheetHeaderTemplateDirective, i4.ActionSheetItemTemplateDirective, i5.ActionSheetContentTemplateDirective, i6.ActionSheetFooterTemplateDirective, i7.ActionSheetTemplateDirective, i8.AppBarComponent, i9.AppBarSectionComponent, i10.AppBarSpacerComponent, i11.BottomNavigationComponent, i12.BottomNavigationItemTemplateDirective, i13.BreadCrumbComponent, i14.BreadCrumbItemTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigationModule, providers: [IconsService, ResizeBatchService], imports: [i1.ActionSheetComponent, i11.BottomNavigationComponent, i13.BreadCrumbComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigationModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_NAVIGATION],
                    exports: [...KENDO_NAVIGATION],
                    providers: [IconsService, ResizeBatchService]
                }]
        }] });
