// @progress/kendo-angular-charts/esm2022/common/prefix-configuration.service.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Injectable, Inject, InjectionToken, NgZone } from '@angular/core';
import { ConfigurationService, Change } from './configuration.service';
import { RootConfigurationService } from './root-configuration.service';
import * as i0 from "@angular/core";
import * as i1 from "./root-configuration.service";
/**
 * @hidden
 */
export const PREFIX = new InjectionToken('configuration prefix');
/**
 * @hidden
 */
export class PrefixConfigurationService extends ConfigurationService {
    rootService;
    prefix;
    constructor(rootService, prefix, ngZone) {
        super(ngZone);
        this.rootService = rootService;
        this.prefix = prefix;
    }
    push(store) {
        this.rootService.notify(new Change(this.prefix, store));
    }
    notify(change) {
        change.key = this.prefix + (change.key ? `.${change.key}` : '');
        this.rootService.notify(change);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PrefixConfigurationService, deps: [{ token: RootConfigurationService }, { token: PREFIX }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PrefixConfigurationService });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PrefixConfigurationService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.RootConfigurationService, decorators: [{
                    type: Inject,
                    args: [RootConfigurationService]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [PREFIX]
                }] }, { type: i0.NgZone }]; } });
