// @progress/kendo-angular-charts/esm2022/sankey/tooltip/link-tooltip-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef, Optional } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Selects a [template](link:site.data.urls.angular['templatesyntax'])
 * within the `<kendo-sankey-tooltip>` component for the
 * [links tooltip](slug:tooltip_sankey#toc-specifying-a-link-tooltip-template).
 * The following context fields are frequently utilized:
 * - `let-source="source"`&mdash;The link source data item..
 * - `let-target="target"`&mdash;The link target data item.
 * - `let-value="value"`&mdash;The link value.
 * Refer to the [`SankeyLinkTooltipTemplateContext`](slug:api_charts_sankeylinktooltiptemplatecontext) for the full list of available fields.
 *
 * @example
 * ```html
 * <kendo-sankey [data]="data">
 *    <kendo-sankey-tooltip [followPointer]="true">
 *        <ng-template kendoSankeyLinkTooltipTemplate let-source="source" let-target="target" let-value="value">
 *            {{ source.label?.text }} - {{ target.label?.text }}: {{ value }}
 *        </ng-template>
 *    </kendo-sankey-tooltip>
 * </kendo-sankey>
 * ```
 */
export class SankeyLinkTooltipTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyLinkTooltipTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: SankeyLinkTooltipTemplateDirective, isStandalone: true, selector: "[kendoSankeyLinkTooltipTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyLinkTooltipTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoSankeyLinkTooltipTemplate]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }]; } });
