// @progress/kendo-angular-charts/esm2022/chart/localization/custom-messages.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { ChartMessages } from './messages';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-l10n";
/**
 * Represents the custom messages component of the Kendo UI for Angular Chart.
 * This component allows you to customize the messages displayed in the Chart, such as the "No data available" message.
 *
 * @example
 * ```html
 * <kendo-chart [data]="data">
 *     <kendo-chart-messages noData='No data'></kendo-chart-messages>
 * </kendo-chart>
 * ```
 */
export class ChartCustomMessagesComponent extends ChartMessages {
    service;
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartCustomMessagesComponent, deps: [{ token: i1.LocalizationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: ChartCustomMessagesComponent, isStandalone: true, selector: "kendo-chart-messages", providers: [
            {
                provide: ChartMessages,
                useExisting: forwardRef(() => ChartCustomMessagesComponent)
            }
        ], usesInheritance: true, ngImport: i0, template: ``, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartCustomMessagesComponent, decorators: [{
            type: Component,
            args: [{
                    providers: [
                        {
                            provide: ChartMessages,
                            useExisting: forwardRef(() => ChartCustomMessagesComponent)
                        }
                    ],
                    selector: 'kendo-chart-messages',
                    template: ``,
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.LocalizationService }]; } });
