// @progress/kendo-angular-dialog/esm2022/window.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { IconsService } from "@progress/kendo-angular-icons";
import { WindowService } from './window/window.service';
import { WindowContainerService } from './window/window-container.service';
import { KENDO_WINDOW } from "./directives";
import * as i0 from "@angular/core";
import * as i1 from "./window/window.component";
import * as i2 from "./window/actions/window-close-action.directive";
import * as i3 from "./window/actions/window-minimize-action.directive";
import * as i4 from "./window/actions/window-maximize-action.directive";
import * as i5 from "./window/actions/window-restore-action.directive";
import * as i6 from "./window/window-titlebar.component";
import * as i7 from "./window/window-container.directive";
import * as i8 from "./dialog/dialog-actions.component";
import * as i9 from "./localization/custom-messages.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmoduleapi']) definition for the Window component.
 *
 * Use the `WindowModule` to include all Window components and directives in your application.
 *
 * @example
 * ```typescript
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { WindowModule } from '@progress/kendo-angular-window';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     bootstrap:    [AppComponent],
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, WindowModule]
 * })
 * export class AppModule {}
 * ```
 */
export class WindowModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WindowModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: WindowModule, imports: [i1.WindowComponent, i2.WindowCloseActionDirective, i3.WindowMinimizeActionDirective, i4.WindowMaximizeActionDirective, i5.WindowRestoreActionDirective, i6.WindowTitleBarComponent, i7.WindowContainerDirective, i8.DialogActionsComponent, i9.CustomMessagesComponent], exports: [i1.WindowComponent, i2.WindowCloseActionDirective, i3.WindowMinimizeActionDirective, i4.WindowMaximizeActionDirective, i5.WindowRestoreActionDirective, i6.WindowTitleBarComponent, i7.WindowContainerDirective, i8.DialogActionsComponent, i9.CustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WindowModule, providers: [WindowContainerService, WindowService, IconsService], imports: [i1.WindowComponent, i2.WindowCloseActionDirective, i3.WindowMinimizeActionDirective, i4.WindowMaximizeActionDirective, i5.WindowRestoreActionDirective, i8.DialogActionsComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: WindowModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_WINDOW],
                    exports: [...KENDO_WINDOW],
                    providers: [WindowContainerService, WindowService, IconsService]
                }]
        }] });
