// @progress/kendo-angular-charts/esm2022/chart/series-defaults/labels.from.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the `from` label configuration of the Chart series.
 */
export class SeriesDefaultsLabelsFromComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the background color of the `from` labels.
     */
    background;
    /**
     * Specifies the border configuration of the `from` labels.
     */
    border;
    /**
     * Specifies the text color of the `from` labels.
     */
    color;
    /**
     * Specifies a function that returns the content of the `from` labels.
     */
    content;
    /**
     * Specifies the font style of the `from` labels.
     * @default '12px sans-serif'
     */
    font;
    /**
     * Specifies the format of the `from` labels. Uses the [`format`]({% slug api_intl_intlservice %}#toc-format) of `IntlService`.
     * @default '{0}'
     */
    format;
    /**
     * Specifies the margin of the `from` labels. A numeric value sets all margins.
     * @default 0
     */
    margin;
    /**
     * Specifies the padding of the `from` labels. A numeric value sets all paddings.
     * @default 0
     */
    padding;
    /**
     * Determines whether the Chart displays the series `from` labels.
     * @default false
     */
    visible;
    constructor(configurationService) {
        super('seriesDefaults.labels.from', configurationService);
        this.configurationService = configurationService;
        this.markAsVisible();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDefaultsLabelsFromComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesDefaultsLabelsFromComponent, isStandalone: true, selector: "kendo-chart-series-defaults-labels-from", inputs: { background: "background", border: "border", color: "color", content: "content", font: "font", format: "format", margin: "margin", padding: "padding", visible: "visible" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDefaultsLabelsFromComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-defaults-labels-from',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], content: [{
                type: Input
            }], font: [{
                type: Input
            }], format: [{
                type: Input
            }], margin: [{
                type: Input
            }], padding: [{
                type: Input
            }], visible: [{
                type: Input
            }] } });
