// @progress/kendo-angular-charts/esm2022/chart/series-defaults/labels.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration of the Chart series labels.
 *
 * @example
 * ```html
 * <kendo-chart>
 *  <kendo-chart-series-defaults>
 *   <kendo-chart-series-defaults-labels format="c">
 *   </kendo-chart-series-defaults-labels>
 *  </kendo-chart-series-defaults>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link SeriesDefaultsLabelsFromComponent} and {@link SeriesDefaultsLabelsToComponent}.
 */
export class SeriesDefaultsLabelsComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the background color of the labels.
     */
    background;
    /**
     * Specifies the border configuration of the labels.
     */
    border;
    /**
     * Specifies the text color of the labels.
     */
    color;
    /**
     * Specifies a function that returns the content of the labels.
     */
    content;
    /**
     * Specifies the font style of the labels.
     * @default '12px sans-serif'
     */
    font;
    /**
     * Specifies the format of the labels. Uses the [`format`]({% slug api_intl_intlservice %}#toc-format) method of `IntlService`.
     * @default '{0}'
     */
    format;
    /**
     * Specifies the margin of the labels. A numeric value sets all margins.
     * @default 0
     */
    margin;
    /**
     * Specifies the padding of the labels. A numeric value sets all paddings.
     * @default 0
     */
    padding;
    /**
     * Determines whether the Chart displays the series labels.
     * @default false
     */
    visible;
    /**
     * Specifies a function that creates a custom visual for the labels.
     */
    visual;
    // These options are also available as child components
    /**
     * Specifies the `from` labels configuration.
     */
    from;
    /**
     * Specifies the `to` labels configuration.
     */
    to;
    constructor(configurationService) {
        super('seriesDefaults.labels', configurationService);
        this.configurationService = configurationService;
        this.markAsVisible();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDefaultsLabelsComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesDefaultsLabelsComponent, isStandalone: true, selector: "kendo-chart-series-defaults-labels", inputs: { background: "background", border: "border", color: "color", content: "content", font: "font", format: "format", margin: "margin", padding: "padding", visible: "visible", visual: "visual", from: "from", to: "to" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDefaultsLabelsComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-defaults-labels',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], content: [{
                type: Input
            }], font: [{
                type: Input
            }], format: [{
                type: Input
            }], margin: [{
                type: Input
            }], padding: [{
                type: Input
            }], visible: [{
                type: Input
            }], visual: [{
                type: Input
            }], from: [{
                type: Input
            }], to: [{
                type: Input
            }] } });
