// @progress/kendo-angular-charts/esm2022/chart/series-defaults/notes.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration of the [`seriesDefaults`]({% slug api_charts_seriesdefaultscomponent %}) notes.
 *
 * @example
 * ```html
 * <kendo-chart>
 * <kendo-chart-series-defaults>
 *   <kendo-chart-series-defaults-notes position="above">
 *   </kendo-chart-series-defaults-notes>
 *  </kendo-chart-series-defaults>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link SeriesDefaultsNotesIconComponent} and {@link SeriesDefaultsNotesLabelComponent}.
 */
export class SeriesDefaultsNotesComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the line configuration for the notes.
     */
    line;
    /**
     * Specifies a function that creates a custom visual for the notes.
     */
    visual;
    // These options are also available as child components
    /**
     * Specifies the icon configuration for the notes.
     */
    icon;
    /**
     * Specifies the label configuration for the notes.
     */
    label;
    constructor(configurationService) {
        super('seriesDefaults.notes', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDefaultsNotesComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesDefaultsNotesComponent, isStandalone: true, selector: "kendo-chart-series-defaults-notes", inputs: { line: "line", visual: "visual", icon: "icon", label: "label" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDefaultsNotesComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-defaults-notes',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { line: [{
                type: Input
            }], visual: [{
                type: Input
            }], icon: [{
                type: Input
            }], label: [{
                type: Input
            }] } });
