// @progress/kendo-angular-charts/esm2022/stock-chart/navigator/series-item/labels.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ConfigurationService } from '../../../common/configuration.service';
import { SeriesLabelsComponent } from '../../../chart/series-item/labels.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../common/configuration.service";
/**
 * Represents the label configuration of the StockChart navigator series.
 *
 * @example
 * ```html
 * <kendo-stockchart>
 *  <kendo-chart-navigator>
 *   <kendo-chart-navigator-series>
 *    <kendo-chart-navigator-series-item type="area" [data]="data" field="value" categoryField="date">
 *    <kendo-chart-navigator-series-item-labels color="peachpuff">
 *    </kendo-chart-navigator-series-item-labels>
 *   </kendo-chart-navigator-series-item>
 *   </kendo-chart-navigator-series>
 *  </kendo-chart-navigator>
 * </kendo-stockchart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link NavigatorSeriesLabelsFromComponent} and {@link NavigatorSeriesLabelsToComponent}.
 */
export class NavigatorSeriesLabelsComponent extends SeriesLabelsComponent {
    configurationService;
    constructor(configurationService) {
        super(configurationService);
        this.configurationService = configurationService;
        this.markAsVisible();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorSeriesLabelsComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: NavigatorSeriesLabelsComponent, isStandalone: true, selector: "kendo-chart-navigator-series-item-labels", usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorSeriesLabelsComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-navigator-series-item-labels',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; } });
