// @progress/kendo-angular-charts/esm2022/stock-chart/navigator/series-item/notes.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ConfigurationService } from '../../../common/configuration.service';
import { SeriesNotesComponent } from '../../../chart/series-item/notes.component';
import * as i0 from "@angular/core";
import * as i1 from "../../../common/configuration.service";
/**
 * Represents the notes configuration of the StockChart navigator series.
 *
 * @example
 * ```html
 * <kendo-stockchart>
 *  <kendo-chart-navigator>
 *   <kendo-chart-navigator-series>
 *    <kendo-chart-navigator-series-item type="area" [data]="data" field="value" categoryField="date">
 *     <kendo-chart-navigator-series-item-notes [visible]="true">
 *     </kendo-chart-navigator-series-item-notes>
 *   </kendo-chart-navigator-series-item>
 *   </kendo-chart-navigator-series>
 *  </kendo-chart-navigator>
 * </kendo-stockchart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link NavigatorSeriesNotesIconComponent} and {@link NavigatorSeriesNotesLabelComponent}.
 */
export class NavigatorSeriesNotesComponent extends SeriesNotesComponent {
    configurationService;
    constructor(configurationService) {
        super(configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorSeriesNotesComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: NavigatorSeriesNotesComponent, isStandalone: true, selector: "kendo-chart-navigator-series-item-notes", usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorSeriesNotesComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-navigator-series-item-notes',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; } });
