// @progress/kendo-angular-charts/esm2022/chart/series-item/error-bars.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration of the Chart series error bars
 * ([see example]({% slug errorbars_chart_charts %})).
 */
export class SeriesErrorBarsComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the color of the error bars.
     */
    color;
    /**
     * Determines whether the caps of the error bars are displayed. By default, the caps are visible.
     * @default true
     */
    endCaps;
    /**
     * Specifies the line configuration of the error bars.
     */
    line;
    /**
     * Specifies the value for the error bars.
     */
    value;
    /**
     * Specifies a function that creates a custom visual for the error bars.
     */
    visual;
    /**
     * Specifies the x-axis value for the error bars.
     */
    xValue;
    /**
     * Specifies the y-axis value for the error bars.
     */
    yValue;
    constructor(configurationService) {
        super('errorBars', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesErrorBarsComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesErrorBarsComponent, isStandalone: true, selector: "kendo-chart-series-item-error-bars", inputs: { color: "color", endCaps: "endCaps", line: "line", value: "value", visual: "visual", xValue: "xValue", yValue: "yValue" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesErrorBarsComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-item-error-bars',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { color: [{
                type: Input
            }], endCaps: [{
                type: Input
            }], line: [{
                type: Input
            }], value: [{
                type: Input
            }], visual: [{
                type: Input
            }], xValue: [{
                type: Input
            }], yValue: [{
                type: Input
            }] } });
