// @progress/kendo-angular-charts/esm2022/chart/series-item/highlight.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the Chart series highlighting configuration options.
 */
export class SeriesHighlightComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the border configuration of the highlight.
     */
    border;
    /**
     * Specifies the color of the highlight.
     */
    color;
    /**
     * Specifies the line configuration of the highlight.
     */
    line;
    /**
     * Specifies the markers configuration of the highlight.
     */
    markers;
    /**
     * Specifies the opacity of the highlight.
     */
    opacity;
    /**
     * Specifies a function that handles the highlight toggle event.
     */
    toggle;
    /**
     * Determines whether the Chart highlights the series when the user hovers over it with the mouse.
     * @default true
     */
    visible;
    /**
     * Specifies a function that creates a custom visual for the highlight.
     */
    visual;
    constructor(configurationService) {
        super('highlight', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesHighlightComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesHighlightComponent, isStandalone: true, selector: "kendo-chart-series-item-highlight", inputs: { border: "border", color: "color", line: "line", markers: "markers", opacity: "opacity", toggle: "toggle", visible: "visible", visual: "visual" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesHighlightComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-item-highlight',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { border: [{
                type: Input
            }], color: [{
                type: Input
            }], line: [{
                type: Input
            }], markers: [{
                type: Input
            }], opacity: [{
                type: Input
            }], toggle: [{
                type: Input
            }], visible: [{
                type: Input
            }], visual: [{
                type: Input
            }] } });
