// @progress/kendo-angular-charts/esm2022/chart/series-item/outliers.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration of the Chart series outliers.
 * Applies to mild outliers.
 * For more information, refer to [`series.outliers`]({% slug api_charts_seriesitemcomponent %}#toc-outliers).
 */
export class SeriesOutliersComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the background color of the outliers.
     */
    background;
    /**
     * Specifies the border configuration of the outliers.
     */
    border;
    /**
     * Specifies the rotation angle of the outliers.
     */
    rotation;
    /**
     * Specifies the marker size in pixels.
     *
     * @default 6
     */
    size;
    /**
     * Specifies the outliers shape.
     *
     * @default 'circle'
     */
    type;
    constructor(configurationService) {
        super('outliers', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesOutliersComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesOutliersComponent, isStandalone: true, selector: "kendo-chart-series-item-outliers", inputs: { background: "background", border: "border", rotation: "rotation", size: "size", type: "type" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesOutliersComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-item-outliers',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], rotation: [{
                type: Input
            }], size: [{
                type: Input
            }], type: [{
                type: Input
            }] } });
