// @progress/kendo-angular-charts/esm2022/chart/series-item/trendline.forecast.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration options of the Chart trendline forecast
 * ([see example]({% slug trendlines_chart_charts %})).
 */
export class SeriesTrendlineForecastComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the number of periods before the series.
     */
    before;
    /**
     * Specifies the number of periods after the series.
     */
    after;
    /**
     * Specifies the series trendline forecast settings.
     *
     * The `forecast` option is supported  when [`series.type`]({% slug api_charts_series %}#toc-type) is set to `"linearTrendline"` and the parent series are either date series, `"scatter"` or `"scatterLine"` series.
     */
    forecast;
    constructor(configurationService) {
        super('forecast', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesTrendlineForecastComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SeriesTrendlineForecastComponent, isStandalone: true, selector: "kendo-chart-series-item-trendline-forecast", inputs: { before: "before", after: "after", forecast: "forecast" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesTrendlineForecastComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-series-item-trendline-forecast',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { before: [{
                type: Input
            }], after: [{
                type: Input
            }], forecast: [{
                type: Input
            }] } });
