// @progress/kendo-angular-buttons/esm2022/dropdownbutton/dropdownbutton.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DROPDOWNBUTTON } from '../directives';
import { PopupService } from '@progress/kendo-angular-popup';
import { IconsService } from '@progress/kendo-angular-icons';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./dropdownbutton.component";
import * as i2 from "../listbutton/button-item-template.directive";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Exports the package module.
 *
 * Required for adding DropDownButton features in NgModule-based Angular applications.
 *
 * The module includes:
 * - `DropDownButtonComponent`&mdash;The component class for the DropDownButton.
 * - `ButtonItemTemplateDirective`&mdash;The directive for the button item template.
 *
 * @example
 * ```typescript
 * import { DropDownButtonModule } from '@progress/kendo-angular-buttons';
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule, DropDownButtonModule],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class DropDownButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DropDownButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DropDownButtonModule, imports: [i1.DropDownButtonComponent, i2.ButtonItemTemplateDirective], exports: [i1.DropDownButtonComponent, i2.ButtonItemTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DropDownButtonModule, providers: [IconsService, PopupService, ResizeBatchService], imports: [i1.DropDownButtonComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DropDownButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_DROPDOWNBUTTON],
                    imports: [...KENDO_DROPDOWNBUTTON],
                    providers: [IconsService, PopupService, ResizeBatchService]
                }]
        }] });
