// @progress/kendo-angular-charts/esm2022/chart/chart-area.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the Kendo UI for Angular Chart Area configuration.
 * Represents the entire visible area of the Chart
 * ([see example]({% slug chartarea_chart_charts %})).
 *
 * @example
 * ```html
 * <kendo-chart>
 *  <kendo-chart-area background="#f0f0f0" [border]="{ color: '#ccc', width: 1 }">
 *  </kendo-chart-area>
 * </kendo-chart>
 * ```
 */
export class ChartAreaComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the background color of the Chart area. Accepts a valid CSS color string, including hex and rgb.
     * @default 'white'
     */
    background;
    /**
     * Specifies the border configuration of the Chart area.
     */
    border;
    /**
     * Specifies the height of the Chart area.
     * @default 400
     */
    height;
    /**
     * Specifies the margin of the Chart area. A numeric value sets all margins.
     * @default 5
     */
    margin;
    /**
     * Specifies the background opacity of the Chart area. By default, the background is opaque.
     * @default 1
     */
    opacity;
    /**
     * Specifies the width of the Chart area.
     * @default 600
     */
    width;
    constructor(configurationService) {
        super('chartArea', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartAreaComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: ChartAreaComponent, isStandalone: true, selector: "kendo-chart-area", inputs: { background: "background", border: "border", height: "height", margin: "margin", opacity: "opacity", width: "width" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartAreaComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-area',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], height: [{
                type: Input
            }], margin: [{
                type: Input
            }], opacity: [{
                type: Input
            }], width: [{
                type: Input
            }] } });
