// @progress/kendo-angular-charts/esm2022/chart/donut-center-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef, Optional } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents a directive which selects a [template](link:site.data.urls.angular['templatesyntax'])
 * within the `<kendo-chart>` component for the
 * [Donut center template](slug:donut_seriestypes_charts#toc-using-the-center-template).
 *
 * @example
 * <kendo-chart>
 *   <ng-template kendoChartDonutCenterTemplate>
 *     <span>Center Content</span>
 *   </ng-template>
 * </kendo-chart>
 */
export class DonutCenterTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DonutCenterTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: DonutCenterTemplateDirective, isStandalone: true, selector: "[kendoChartDonutCenterTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DonutCenterTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoChartDonutCenterTemplate]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }]; } });
