// @progress/kendo-angular-charts/esm2022/chart/panes.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ContentChildren, QueryList } from '@angular/core';
import { CollectionComponent } from '../common/collection.component';
import { CollectionService } from '../common/collection.service';
import { ConfigurationService } from '../common/configuration.service';
import { PaneComponent } from './pane.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
import * as i2 from "../common/collection.service";
/**
 * Represents collection of one or more pane configuration components.
 *
 * @example
 * ```html
 * <kendo-chart>
 *   <kendo-chart-panes>
 *     <kendo-chart-pane name="topPane"></kendo-chart-pane>
 *     <kendo-chart-pane name="bottomPane"></kendo-chart-pane>
 *   </kendo-chart-panes>
 *   ...
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link PaneComponent}.
 */
export class PanesComponent extends CollectionComponent {
    configurationService;
    collectionService;
    children;
    constructor(configurationService, collectionService) {
        super('panes', configurationService, collectionService);
        this.configurationService = configurationService;
        this.collectionService = collectionService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PanesComponent, deps: [{ token: i1.ConfigurationService }, { token: i2.CollectionService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: PanesComponent, isStandalone: true, selector: "kendo-chart-panes", providers: [CollectionService], queries: [{ propertyName: "children", predicate: PaneComponent }], usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PanesComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [CollectionService],
                    selector: 'kendo-chart-panes',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }, { type: i2.CollectionService }]; }, propDecorators: { children: [{
                type: ContentChildren,
                args: [PaneComponent]
            }] } });
