// @progress/kendo-angular-charts/esm2022/chart/pane-defaults.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the default options for all Chart panes
 * ([see example](slug:panes_chart_charts#toc-default-settings)).
 *
 * @example
 * ```html
 * <kendo-chart>
 *  <kendo-chart-pane-defaults background="#f0f0f0" [border]="{ color: '#ccc', width: 1 }">
 *   <kendo-chart-pane-defaults-title [visible]="true"></kendo-chart-pane-defaults-title>
 *  </kendo-chart-pane-defaults>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link PaneDefaultsTitleComponent}.
 */
export class PaneDefaultsComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the background color of the pane.
     */
    background;
    /**
     * Specifies the border configuration of the pane.
     */
    border;
    /**
     * Determines whether the pane clips its content.
     */
    clip;
    /**
     * Specifies the height of the pane.
     */
    height;
    /**
     * Specifies the margin of the pane. A numeric value sets all margins.
     * @default 0
     */
    margin;
    /**
     * Specifies the padding of the pane. A numeric value sets all paddings.
     * @default 0
     */
    padding;
    /**
     * Specifies the title configuration of the pane.
     */
    title;
    constructor(configurationService) {
        super('paneDefaults', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PaneDefaultsComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: PaneDefaultsComponent, isStandalone: true, selector: "kendo-chart-pane-defaults", inputs: { background: "background", border: "border", clip: "clip", height: "height", margin: "margin", padding: "padding", title: "title" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PaneDefaultsComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-pane-defaults',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], clip: [{
                type: Input
            }], height: [{
                type: Input
            }], margin: [{
                type: Input
            }], padding: [{
                type: Input
            }], title: [{
                type: Input
            }] } });
