// @progress/kendo-angular-charts/esm2022/chart/plot-area.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the Kendo UI for Angular Chart Plot Area component
 * ([see example]({% slug plotarea_chart_charts %})).
 * The plot area is the area which displays the series.
 */
export class PlotAreaComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the background color of the Chart plot area.
     * Accepts a valid CSS color string, including HEX and RGB.
     * @default 'white'
     */
    background;
    /**
     * Specifies the border configuration of the plot area.
     */
    border;
    /**
     * Specifies the margin of the plot area. A numeric value sets all margins.
     * @default 5
     */
    margin;
    /**
     * Specifies the background opacity of the plot area. By default, the background is opaque.
     * @default 1
     */
    opacity;
    /**
     * Specifies the padding of the plot area. A numeric value sets all paddings.
     * Use proportional padding for the Pie, Donut, Radar, and Polar Charts based on the Chart size.
     * @default 5
     */
    padding;
    constructor(configurationService) {
        super('plotArea', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PlotAreaComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: PlotAreaComponent, isStandalone: true, selector: "kendo-chart-plot-area", inputs: { background: "background", border: "border", margin: "margin", opacity: "opacity", padding: "padding" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: PlotAreaComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-plot-area',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], margin: [{
                type: Input
            }], opacity: [{
                type: Input
            }], padding: [{
                type: Input
            }] } });
