// @progress/kendo-angular-charts/esm2022/chart/series-drilldown-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef, Optional } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents the Kendo UI for Angular Chart Series Drilldown Template directive.
 * Use this directive within the `<kendo-chart-series-item>` component to create nested
 * [Drilldown](slug:drilldown_chart_charts) series.
 *
 * When you click on a data point with drilldown data,
 * the Chart creates a new series based on this template.
 *
 * The template context passes through the following fields:
 * - `drilldownValue`&mdash;The drilldown field value.
 * - `point`&mdash;The series point to drill down to.
 * - `series`&mdash;The series to drill down to.
 */
export class SeriesDrilldownTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDrilldownTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: SeriesDrilldownTemplateDirective, isStandalone: true, selector: "[kendoChartDrilldownSeries]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SeriesDrilldownTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoChartDrilldownSeries]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }]; } });
