// @progress/kendo-angular-charts/esm2022/chart/subtitle.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the Kendo UI for Angular Chart Subtitle component
 * ([see example]({% slug title_chart_charts %})).
 *
 * @example
 * ```html
 * <kendo-chart [categoryAxis]="{ categories: categories }">
 *     <kendo-chart-subtitle text="/GDP annual %/"></kendo-chart-subtitle>
 * </kendo-chart>
 * ```
 */
export class SubtitleComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the horizontal alignment of the subtitle.
     */
    align;
    /**
     * Specifies the background color of the subtitle. Accepts a valid CSS color string, including HEX and RGB.
     *
     * @default 'white'
     */
    background;
    /**
     * Specifies the border configuration of the subtitle.
     */
    border;
    /**
     * Specifies the color of the subtitle text.
     */
    color;
    /**
     * Specifies the font of the subtitle.
     *
     * @default '12px sans-serif'
     */
    font;
    /**
     * Specifies the margin configuration of the subtitle.
     */
    margin;
    /**
     * Specifies the padding of the subtitle. A numeric value sets all paddings.
     *
     * @default 5
     */
    padding;
    /**
     * Specifies the position of the subtitle.
     *
     * @default 'top'
     */
    position;
    /**
     * Specifies the text content of the subtitle.
     */
    text;
    /**
     * Determines whether the Chart displays the subtitle.
     *
     * @default true
     */
    visible;
    constructor(configurationService) {
        super('subtitle', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SubtitleComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SubtitleComponent, isStandalone: true, selector: "kendo-chart-subtitle", inputs: { align: "align", background: "background", border: "border", color: "color", font: "font", margin: "margin", padding: "padding", position: "position", text: "text", visible: "visible" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SubtitleComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-subtitle',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { align: [{
                type: Input
            }], background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], font: [{
                type: Input
            }], margin: [{
                type: Input
            }], padding: [{
                type: Input
            }], position: [{
                type: Input
            }], text: [{
                type: Input
            }], visible: [{
                type: Input
            }] } });
