// @progress/kendo-angular-charts/esm2022/chart/title.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the Kendo UI for Angular Chart Title configuration
 * ([see example]({% slug title_chart_charts %})).
 *
 * @example
 * ```html
 * <kendo-chart [categoryAxis]="{ categories: categories }">
 *     <kendo-chart-title text="Title Text" [padding]="{bottom: 20}"></kendo-chart-title>
 *     ...
 * </kendo-chart>
 * ```
 */
export class TitleComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the alignment of the title.
     * The alignment options for text include `center` for middle alignment, `left` for left alignment, and `right` for right alignment, allowing for flexible positioning of the title.
     *
     * @default 'center'
     */
    align;
    /**
     * Specifies the background color of the title. Accepts a valid CSS color string, including HEX and RGB.
     *
     * @default 'white'
     */
    background;
    /**
     * Specifies the border configuration of the title.
     */
    border;
    /**
     * Specifies the text color of the title.
     */
    color;
    /**
     * Specifies the font of the title.
     *
     * @default '16px sans-serif'
     */
    font;
    /**
     * Specifies the margin of the title. A numeric value sets all margins.
     */
    margin;
    /**
     * Specifies the padding of the title. A numeric value sets all paddings.
     *
     * @default 5
     */
    padding;
    /**
     * Specifies the position of the title.
     * The positioning options for titles include `bottom` for positioning at the bottom and `top` for positioning at the top, providing flexibility in title placement.
     *
     * @default 'top'
     */
    position;
    /**
     * Specifies the text of the title.
     */
    text;
    /**
     * Specifies the description of the title.
     */
    description;
    /**
     * Determines whether the Chart displays the title.
     *
     * @default true
     */
    visible;
    constructor(configurationService) {
        super('title', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TitleComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: TitleComponent, isStandalone: true, selector: "kendo-chart-title", inputs: { align: "align", background: "background", border: "border", color: "color", font: "font", margin: "margin", padding: "padding", position: "position", text: "text", description: "description", visible: "visible" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TitleComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-title',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { align: [{
                type: Input
            }], background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], font: [{
                type: Input
            }], margin: [{
                type: Input
            }], padding: [{
                type: Input
            }], position: [{
                type: Input
            }], text: [{
                type: Input
            }], description: [{
                type: Input
            }], visible: [{
                type: Input
            }] } });
