// @progress/kendo-angular-charts/esm2022/chart/y-axis.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ContentChildren, QueryList } from '@angular/core';
import { CollectionComponent } from '../common/collection.component';
import { CollectionService } from '../common/collection.service';
import { ConfigurationService } from '../common/configuration.service';
import { YAxisItemComponent } from './y-axis-item.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
import * as i2 from "../common/collection.service";
/**
 * Represents a collection of one or more Y-axis configuration components.
 *
 * @example
 * ```html
 * <kendo-chart>
 *     <kendo-chart-y-axis>
 *         <kendo-chart-y-axis-item
 *           [background]="'rgba(100, 100, 100, 0.2)'"
 *           color="red"
 *           [notes]="notesConfig"
 *           [crosshair]="crosshairConfig">
 *         </kendo-chart-y-axis-item>
 *         <kendo-chart-y-axis-item name="secondAxis">
 *         </kendo-chart-y-axis-item>
 *     </kendo-chart-y-axis>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link YAxisItemComponent}.
 */
export class YAxisComponent extends CollectionComponent {
    configurationService;
    collectionService;
    children;
    constructor(configurationService, collectionService) {
        super('yAxis', configurationService, collectionService);
        this.configurationService = configurationService;
        this.collectionService = collectionService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: YAxisComponent, deps: [{ token: i1.ConfigurationService }, { token: i2.CollectionService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: YAxisComponent, isStandalone: true, selector: "kendo-chart-y-axis", providers: [CollectionService], queries: [{ propertyName: "children", predicate: YAxisItemComponent }], usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: YAxisComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [CollectionService],
                    selector: 'kendo-chart-y-axis',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }, { type: i2.CollectionService }]; }, propDecorators: { children: [{
                type: ContentChildren,
                args: [YAxisItemComponent]
            }] } });
