// @progress/kendo-angular-charts/esm2022/chart/y-axis-item.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { CollectionItemComponent } from '../common/collection-item.component';
import { CollectionService } from '../common/collection.service';
import { ConfigurationService } from '../common/configuration.service';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
import * as i2 from "../common/collection.service";
/**
 * Represents the Kendo UI for Angular Chart Y Axis Item component
 * ([see example]({% slug api_charts_yaxiscomponent %})).
 *
 * @example
 * ```html
 * <kendo-chart>
 *  <kendo-chart-y-axis>
 *   <kendo-chart-y-axis-item color="indigo" [line]="{ width: 2 }">
 *   </kendo-chart-y-axis-item>
 * </kendo-chart-y-axis>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link YAxisCrosshairComponent}, {@link YAxisLabelsComponent}, {@link YAxisNotesComponent}, and {@link YAxisTitleComponent}.
 */
export class YAxisItemComponent extends CollectionItemComponent {
    configurationService;
    collectionService;
    /**
     * Specifies the value or array of values at which the axis crosses with another axis.
     */
    axisCrossingValue;
    /**
     * Specifies the background color of the axis.
     */
    background;
    /**
     * Specifies the base time unit for the axis.
     */
    baseUnit;
    /**
     * Specifies the categories for the axis.
     */
    categories;
    /**
     * Specifies the color of the axis.
     */
    color;
    /**
     * Specifies the configuration of the axis line.
     */
    line;
    /**
     * Specifies the configuration of the major grid lines.
     */
    majorGridLines;
    /**
     * Specifies the configuration of the major ticks.
     */
    majorTicks;
    /**
     * Specifies the interval between major divisions.
     */
    majorUnit;
    /**
     * Specifies the maximum value of the axis.
     */
    max;
    /**
     * Specifies the minimum value of the axis.
     */
    min;
    /**
     * Specifies the configuration of the minor grid lines.
     */
    minorGridLines;
    /**
     * Specifies the configuration of the minor ticks.
     */
    minorTicks;
    /**
     * Specifies the interval between minor divisions.
     */
    minorUnit;
    /**
     * Specifies the unique name of the axis.
     */
    name;
    /**
     * Determines whether the Chart prevents the automatic axis range from snapping to zero.
     * Set to `false` to force the automatic axis range to snap to zero.
     * @default false
     */
    narrowRange;
    /**
     * Specifies the name of the pane that the axis renders in.
     */
    pane;
    /**
     * Specifies the plot bands configuration.
     */
    plotBands;
    /**
     * Determines whether the value axis direction is reversed.
     * By default, the values increase from left to right and from bottom to top.
     * @default false
     */
    reverse;
    /**
     * Specifies the axis type.
     *
     * The options are `numeric` for a numeric axis, `date` for a specialized axis for displaying chronological data, and `log` for a logarithmic axis.
     *
     * When the series Y value is of the `date` type, the Chart automatically switches to a date axis.
     * To avoid this behavior, set the `type`.
     * @default 'numeric'
     */
    type;
    /**
     * Determines whether the Chart displays the Y axis. By default, the Y axis is visible.
     * @default true
     */
    visible;
    // These options are also available as child components
    /**
     * Specifies the configuration of the crosshair.
     */
    crosshair;
    /**
     * Specifies the configuration of the labels.
     */
    labels;
    /**
     * Specifies the configuration of the notes.
     */
    notes;
    /**
     * Specifies the configuration of the title.
     */
    title;
    constructor(configurationService, collectionService) {
        super(configurationService, collectionService);
        this.configurationService = configurationService;
        this.collectionService = collectionService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: YAxisItemComponent, deps: [{ token: i1.ConfigurationService }, { token: i2.CollectionService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: YAxisItemComponent, isStandalone: true, selector: "kendo-chart-y-axis-item", inputs: { axisCrossingValue: "axisCrossingValue", background: "background", baseUnit: "baseUnit", categories: "categories", color: "color", line: "line", majorGridLines: "majorGridLines", majorTicks: "majorTicks", majorUnit: "majorUnit", max: "max", min: "min", minorGridLines: "minorGridLines", minorTicks: "minorTicks", minorUnit: "minorUnit", name: "name", narrowRange: "narrowRange", pane: "pane", plotBands: "plotBands", reverse: "reverse", type: "type", visible: "visible", crosshair: "crosshair", labels: "labels", notes: "notes", title: "title" }, providers: [ConfigurationService], usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: YAxisItemComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [ConfigurationService],
                    selector: 'kendo-chart-y-axis-item',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }, { type: i2.CollectionService }]; }, propDecorators: { axisCrossingValue: [{
                type: Input
            }], background: [{
                type: Input
            }], baseUnit: [{
                type: Input
            }], categories: [{
                type: Input
            }], color: [{
                type: Input
            }], line: [{
                type: Input
            }], majorGridLines: [{
                type: Input
            }], majorTicks: [{
                type: Input
            }], majorUnit: [{
                type: Input
            }], max: [{
                type: Input
            }], min: [{
                type: Input
            }], minorGridLines: [{
                type: Input
            }], minorTicks: [{
                type: Input
            }], minorUnit: [{
                type: Input
            }], name: [{
                type: Input
            }], narrowRange: [{
                type: Input
            }], pane: [{
                type: Input
            }], plotBands: [{
                type: Input
            }], reverse: [{
                type: Input
            }], type: [{
                type: Input
            }], visible: [{
                type: Input
            }], crosshair: [{
                type: Input
            }], labels: [{
                type: Input
            }], notes: [{
                type: Input
            }], title: [{
                type: Input
            }] } });
