// @progress/kendo-angular-charts/esm2022/chart/legend/item.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the Kendo UI for Angular Chart Legend Item configuration.
 *
 * @example
 *
 * ```html
 * <kendo-chart>
 *   <kendo-chart-legend>
 *      <kendo-chart-legend-item cursor="crosshair"></kendo-chart-legend-item>
 *   </kendo-chart-legend>
 * </kendo-chart>
 * ```
 */
export class LegendItemComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the cursor style for legend items.
     */
    cursor;
    /**
     * Specifies the type of the legend item.
     */
    type;
    /**
     * Specifies the line configuration for the legend item.
     */
    line;
    /**
     * Specifies the area configuration for the legend item.
     */
    area;
    /**
     * Specifies the markers configuration for the legend item.
     */
    markers;
    /**
     * Specifies the highlight configuration for the legend item.
     */
    highlight;
    /**
     * Specifies a function that creates a custom visual for the legend item.
     */
    visual;
    constructor(configurationService) {
        super('legend.item', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LegendItemComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: LegendItemComponent, isStandalone: true, selector: "kendo-chart-legend-item", inputs: { cursor: "cursor", type: "type", line: "line", area: "area", markers: "markers", highlight: "highlight", visual: "visual" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: LegendItemComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-legend-item',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { cursor: [{
                type: Input
            }], type: [{
                type: Input
            }], line: [{
                type: Input
            }], area: [{
                type: Input
            }], markers: [{
                type: Input
            }], highlight: [{
                type: Input
            }], visual: [{
                type: Input
            }] } });
