// @progress/kendo-angular-charts/esm2022/chart/category-axis-item/crosshair.tooltip.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration options of the crosshair tooltip for the category axis of the Chart ([see example]({% slug crosshairs_chart_charts %})).
 * The crosshair tooltip is displayed when the `visible` option is set to `true`.
 */
export class CategoryAxisCrosshairTooltipComponent extends SettingsComponent {
    configurationService;
    /**
     * Sets the background color of the tooltip.
     */
    background;
    /**
     * Specifies the border configuration of the tooltip.
     */
    border;
    /**
     * Defines the text color of the tooltip.
     */
    color;
    /**
     * Sets the font of the tooltip.
     * @default '12px sans-serif'
     */
    font;
    /**
     * Specifies the format used to display the tooltip. Uses the [`format`]({% slug api_intl_intlservice %}#toc-format) method of IntlService.
     * Contains one placeholder (`"{0}"`) which represents the category value.
     * @default '{0}'
     */
    format;
    /**
     * Specifies the padding of the crosshair tooltip. A numeric value sets all paddings.
     * @default 0
     */
    padding;
    /**
     * Determines whether the Chart displays the category axis crosshair tooltip.
     * By default, the category axis crosshair tooltip is not visible.
     * @default false
     */
    visible;
    constructor(configurationService) {
        super('crosshair.tooltip', configurationService);
        this.configurationService = configurationService;
        this.markAsVisible();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisCrosshairTooltipComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: CategoryAxisCrosshairTooltipComponent, isStandalone: true, selector: "kendo-chart-category-axis-item-crosshair-tooltip", inputs: { background: "background", border: "border", color: "color", font: "font", format: "format", padding: "padding", visible: "visible" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisCrosshairTooltipComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-category-axis-item-crosshair-tooltip',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], font: [{
                type: Input
            }], format: [{
                type: Input
            }], padding: [{
                type: Input
            }], visible: [{
                type: Input
            }] } });
