// @progress/kendo-angular-charts/esm2022/chart/category-axis-item/crosshair.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Configures the crosshair for the category axis in the Chart ([see example]({% slug crosshairs_chart_charts %})).
 *
 * @remarks
 * Supported children components are: {@link CategoryAxisCrosshairTooltipComponent}.
 */
export class CategoryAxisCrosshairComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the color of the crosshair.
     */
    color;
    /**
     * Defines the dash type of the crosshair.
     * @default 'solid'
     */
    dashType;
    /**
     * Specifies the opacity of the crosshair. By default, the crosshair is opaque.
     * @default 1
     */
    opacity;
    /**
     * Determines whether the Chart displays the category axis crosshair.
     * By default, the category axis crosshair is not visible.
     * @default false
     */
    visible;
    /**
     * Sets the width of the crosshair in pixels.
     * @default 1
     */
    width;
    // These options are also available as child components
    /**
     * Configures the tooltip configuration of the crosshair.
     */
    tooltip;
    constructor(configurationService) {
        super('crosshair', configurationService);
        this.configurationService = configurationService;
        this.markAsVisible();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisCrosshairComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: CategoryAxisCrosshairComponent, isStandalone: true, selector: "kendo-chart-category-axis-item-crosshair", inputs: { color: "color", dashType: "dashType", opacity: "opacity", visible: "visible", width: "width", tooltip: "tooltip" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisCrosshairComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-category-axis-item-crosshair',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { color: [{
                type: Input
            }], dashType: [{
                type: Input
            }], opacity: [{
                type: Input
            }], visible: [{
                type: Input
            }], width: [{
                type: Input
            }], tooltip: [{
                type: Input
            }] } });
