// @progress/kendo-angular-charts/esm2022/chart/category-axis-item/notes.label.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the label configuration for the category axis notes in the Chart.
 */
export class CategoryAxisNotesLabelComponent extends SettingsComponent {
    configurationService;
    /**
     * Sets the background color of the label.
     */
    background;
    /**
     * Configures the border of the label.
     */
    border;
    /**
     * Sets the text color of the label.
     */
    color;
    /**
     * Provides a function that creates the content of the label.
     */
    content;
    /**
     * Specifies the font style of the label.
     * @default '12px, sans-serif'
     */
    font;
    /**
     * Specifies the format for displaying the notes label. Uses the [`format`]({% slug api_intl_intlservice %}#toc-format) method of IntlService.
     * Contains one placeholder (`"{0}"`) which represents the category value.
     * @default '{0}'
     */
    format;
    /**
     * Specifies the position of the label.
     * @default "inside"
     */
    position;
    /**
     * Specifies the rotation angle of the label. By default, the label is not rotated.
     * @default 0
     */
    rotation;
    /**
     * Determines whether the Chart displays the category notes label.
     * By default, the category notes label is visible.
     * @default true
     */
    visible;
    constructor(configurationService) {
        super('notes.label', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisNotesLabelComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: CategoryAxisNotesLabelComponent, isStandalone: true, selector: "kendo-chart-category-axis-item-notes-label", inputs: { background: "background", border: "border", color: "color", content: "content", font: "font", format: "format", position: "position", rotation: "rotation", visible: "visible" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisNotesLabelComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-category-axis-item-notes-label',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], content: [{
                type: Input
            }], font: [{
                type: Input
            }], format: [{
                type: Input
            }], position: [{
                type: Input
            }], rotation: [{
                type: Input
            }], visible: [{
                type: Input
            }] } });
