// @progress/kendo-angular-charts/esm2022/chart/category-axis-item/select.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Configures the selection for the category axis in the Chart. When set, axis selection is enabled and the range is index-based, starting from zero. You can specify the lower and upper boundaries, minimum and maximum values, and mousewheel configuration. If the categories are dates, specify the range with date values.
 */
export class CategoryAxisSelectComponent extends SettingsComponent {
    configurationService;
    /**
     * Sets the lower boundary of the selected range.
     */
    from;
    /**
     * Sets the maximum value for the selection.
     */
    max;
    /**
     * Sets the minimum value for the selection.
     */
    min;
    /**
     * Specifies the mousewheel configuration for the selection.
     */
    mousewheel;
    /**
     * Specifies the upper boundary of the selected range.
     */
    to;
    constructor(configurationService) {
        super('select', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisSelectComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: CategoryAxisSelectComponent, isStandalone: true, selector: "kendo-chart-category-axis-item-select", inputs: { from: "from", max: "max", min: "min", mousewheel: "mousewheel", to: "to" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CategoryAxisSelectComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-category-axis-item-select',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { from: [{
                type: Input
            }], max: [{
                type: Input
            }], min: [{
                type: Input
            }], mousewheel: [{
                type: Input
            }], to: [{
                type: Input
            }] } });
