// @progress/kendo-angular-dropdowns/esm2022/dropdownlist/dropdownlist.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_DROPDOWNLIST } from '../directives';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { AdaptiveService } from '@progress/kendo-angular-utils';
import * as i0 from "@angular/core";
import * as i1 from "./dropdownlist.component";
import * as i2 from "../common/templates/fixed-group-template.directive";
import * as i3 from "../common/templates/footer-template.directive";
import * as i4 from "../common/templates/group-template.directive";
import * as i5 from "../common/templates/header-template.directive";
import * as i6 from "../common/templates/item-template.directive";
import * as i7 from "../common/templates/value-template.directive";
import * as i8 from "../common/templates/no-data-template.directive";
import * as i9 from "../common/filtering/filter.directive";
import * as i10 from "../common/localization/custom-messages.component";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Required for adding all DropDownList features in NgModule-based Angular applications.
 *
 * ```typescript
 * import { DropDownListModule } from '@progress/kendo-angular-dropdowns';
 * @NgModule({
 *  imports: [DropDownListModule],
 *  declarations: [AppComponent],
 *  bootstrap: [AppComponent]
 * })
 * ```
 *
 * The package exports:
 * - `DropDownListComponent`&mdash;The `DropDownListComponent` class.
 * - `ItemTemplateDirective`&mdash;The `ItemTemplateDirective` directive.
 * - `ValueTemplateDirective`&mdash;The `ValueTemplateDirective` directive.
 * - `HeaderTemplateDirective`&mdash;The `HeaderTemplateDirective` directive.
 * - `FooterTemplateDirective`&mdash;The `FooterTemplateDirective` directive.
 * - `NoDataTemplateDirective`&mdash;The `NoDataTemplateDirective` directive.
 * - `GroupTemplateDirective`&mdash;The `GroupTemplateDirective` directive.
 * - `FixedGroupTemplateDirective`&mdash;The `FixedGroupTemplateDirective` directive.
 * - `FilterDirective`&mdash;The `FilterDirective` directive.
 * - `CustomMessagesComponent`&mdash;The `CustomMessagesComponent` component.
 *
 * @example
 * ```typescript
 * import { DropDownListModule } from '@progress/kendo-angular-dropdowns';
 *
 * @NgModule({
 *   imports: [DropDownListModule],
 * })
 * export class AppModule {}
 * ```
 */
export class DropDownListModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DropDownListModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: DropDownListModule, imports: [i1.DropDownListComponent, i2.FixedGroupTemplateDirective, i3.FooterTemplateDirective, i4.GroupTemplateDirective, i5.HeaderTemplateDirective, i6.ItemTemplateDirective, i7.ValueTemplateDirective, i8.NoDataTemplateDirective, i9.FilterDirective, i10.CustomMessagesComponent], exports: [i1.DropDownListComponent, i2.FixedGroupTemplateDirective, i3.FooterTemplateDirective, i4.GroupTemplateDirective, i5.HeaderTemplateDirective, i6.ItemTemplateDirective, i7.ValueTemplateDirective, i8.NoDataTemplateDirective, i9.FilterDirective, i10.CustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DropDownListModule, providers: [
            PopupService,
            ResizeBatchService,
            IconsService,
            AdaptiveService
        ], imports: [i1.DropDownListComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: DropDownListModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_DROPDOWNLIST],
                    exports: [...KENDO_DROPDOWNLIST],
                    providers: [
                        PopupService,
                        ResizeBatchService,
                        IconsService,
                        AdaptiveService
                    ]
                }]
        }] });
