// @progress/kendo-angular-dropdowns/esm2022/autocomplete/autocomplete.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_AUTOCOMPLETE } from '../directives';
import { KENDO_ADORNMENTS, ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { AdaptiveService } from '@progress/kendo-angular-utils';
import * as i0 from "@angular/core";
import * as i1 from "./autocomplete.component";
import * as i2 from "../common/templates/fixed-group-template.directive";
import * as i3 from "../common/templates/footer-template.directive";
import * as i4 from "../common/templates/group-template.directive";
import * as i5 from "../common/templates/header-template.directive";
import * as i6 from "../common/templates/item-template.directive";
import * as i7 from "../common/templates/no-data-template.directive";
import * as i8 from "../common/filtering/filter.directive";
import * as i9 from "../common/localization/custom-messages.component";
import * as i10 from "@progress/kendo-angular-common";
// IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Required for adding all AutoComplete features in NgModule-based Angular applications.
 *
 *  ```typescript
 *  import { AutoCompleteModule } from '@progress/kendo-angular-dropdowns';
 *  @NgModule({
 *  imports: [AutoCompleteModule],
 *  declarations: [AppComponent],
 *  bootstrap: [AppComponent]
 *  })
 *  ```
 *
 * The package exports:
 * - `AutoCompleteComponent`&mdash;The AutoComplete component class.
 * - `ItemTemplateDirective`&mdash;The item template directive.
 * - `HeaderTemplateDirective`&mdash;The header template directive.
 * - `FooterTemplateDirective`&mdash;The footer template directive.
 * - `NoDataTemplateDirective`&mdash;The no data template directive.
 * - `FixedGroupTemplateDirective`&mdash;The fixed group template directive.
 * - `GroupTemplateDirective`&mdash;The group template directive.
 * - `FilterDirective`&mdash;The filter directive.
 * - `CustomMessagesComponent`&mdash;The custom messages component.
 * - `SuffixTemplateDirective`&mdash;The suffix template directive.
 * - `PrefixTemplateDirective`&mdash;The prefix template directive.
 */
export class AutoCompleteModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AutoCompleteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: AutoCompleteModule, imports: [i1.AutoCompleteComponent, i2.FixedGroupTemplateDirective, i3.FooterTemplateDirective, i4.GroupTemplateDirective, i5.HeaderTemplateDirective, i6.ItemTemplateDirective, i7.NoDataTemplateDirective, i8.FilterDirective, i9.CustomMessagesComponent, i10.SuffixTemplateDirective, i10.PrefixTemplateDirective, i10.SeparatorComponent, i10.PrefixTemplateDirective, i10.SuffixTemplateDirective, i10.SeparatorComponent], exports: [i1.AutoCompleteComponent, i2.FixedGroupTemplateDirective, i3.FooterTemplateDirective, i4.GroupTemplateDirective, i5.HeaderTemplateDirective, i6.ItemTemplateDirective, i7.NoDataTemplateDirective, i8.FilterDirective, i9.CustomMessagesComponent, i10.SuffixTemplateDirective, i10.PrefixTemplateDirective, i10.SeparatorComponent, i10.PrefixTemplateDirective, i10.SuffixTemplateDirective, i10.SeparatorComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AutoCompleteModule, providers: [
            PopupService,
            ResizeBatchService,
            IconsService,
            AdaptiveService
        ], imports: [i1.AutoCompleteComponent, i10.SeparatorComponent, i10.SeparatorComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: AutoCompleteModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_AUTOCOMPLETE, ...KENDO_ADORNMENTS],
                    exports: [...KENDO_AUTOCOMPLETE, ...KENDO_ADORNMENTS],
                    providers: [
                        PopupService,
                        ResizeBatchService,
                        IconsService,
                        AdaptiveService
                    ]
                }]
        }] });
