// @progress/kendo-angular-common/esm2022/directives.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { PrefixTemplateDirective } from "./adornments/prefix.directive";
import { SeparatorComponent } from "./adornments/separator.component";
import { SuffixTemplateDirective } from "./adornments/suffix.directive";
import { DraggableDirective } from "./draggable/draggable.directive";
import { EventsOutsideAngularDirective } from "./events/events-outside-angular.directive";
import { ResizeSensorComponent } from "./resize-sensor/resize-sensor.component";
import { TemplateContextDirective } from "./template-context/template-context.directive";
import { ToggleButtonTabStopDirective } from "./toggle-button-tab-stop/toggle-button-tab-stop.directive";
import { WatermarkOverlayComponent } from "./watermark/watermark.component";
/**
 * @hidden
 */
export const KENDO_ADORNMENTS = [
    PrefixTemplateDirective,
    SuffixTemplateDirective,
    SeparatorComponent
];
/**
 * @hidden
 */
export const KENDO_COMMON = [
    ...KENDO_ADORNMENTS,
    DraggableDirective,
    EventsOutsideAngularDirective,
    ResizeSensorComponent,
    ToggleButtonTabStopDirective,
    WatermarkOverlayComponent,
];
/**
 * @hidden
 */
export const KENDO_DRAGGABLE = [
    DraggableDirective
];
/**
 * @hidden
 */
export const KENDO_EVENTS = [
    EventsOutsideAngularDirective
];
/**
 * @hidden
 */
export const KENDO_RESIZESENSOR = [
    ResizeSensorComponent
];
/**
 * @hidden
 */
export const KENDO_TOGGLEBUTTONTABSTOP = [
    ToggleButtonTabStopDirective
];
/**
 * @hidden
 */
export const KENDO_WATERMARK = [
    WatermarkOverlayComponent
];
/**
 * @hidden
 */
export const KENDO_TEMPLATE_CONTEXT = [
    TemplateContextDirective
];
