// @progress/kendo-angular-charts/esm2022/chart/tooltip/tooltip-template-point.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
/**
 * Represents the point that is passed to the tooltip template.
 */
export class TooltipTemplatePoint {
    /**
     * Specifies the point value.
     */
    value;
    /**
     * Specifies the point category. Available only for the Categorical series.
     */
    category;
    /**
     * Specifies the point category index. Available only for the Categorical series.
     */
    categoryIndex;
    /**
     * Specifies the point series options.
     */
    series;
    /**
     * Specifies the point `dataItem`.
     */
    dataItem;
    /**
     * Specifies the point value represented as a percentage value. Available only for the Donut, Pie, and 100% Stacked charts.
     */
    percentage;
    /**
     * Specifies the sum of point values since the last `"runningTotal"` summary point. Available for the Waterfall series.
     */
    runningTotal;
    /**
     * Specifies the sum of all previous series values. Available for the Waterfall series.
     */
    total;
    /**
     * Specifies the low value of the point error bar. Available for the Categorical series with error bars.
     */
    low;
    /**
     * Specifies the high value of the point error bar. Available for the Categorical series with error bars.
     */
    high;
    /**
     * Specifies the x low value of the point error bar. Available for the Scatter series with error bars.
     */
    xLow;
    /**
     * Specifies the x high value of the point error bar. Available for the Scatter series with error bars.
     */
    xHigh;
    /**
     * Specifies the y low value of the point error bar. Available for the Scatter series with error bars.
     */
    yLow;
    /**
     * Specifies the y high value of the point error bar. Available for the Scatter series with error bars.
     */
    yHigh;
    /**
     * @hidden
     */
    template;
    /**
     * @hidden
     */
    point;
    /**
     * @hidden
     */
    format;
    /**
     * @hidden
     */
    constructor(point, format, template) {
        this.value = point.value;
        this.category = point.category;
        this.categoryIndex = point.categoryIx;
        this.series = point.series;
        this.dataItem = point.dataItem;
        this.percentage = point.percentage;
        this.runningTotal = point.runningTotal;
        this.total = point.total;
        this.low = point.low;
        this.high = point.high;
        this.xLow = point.xLow;
        this.xHigh = point.xHigh;
        this.yLow = point.yLow;
        this.yHigh = point.yHigh;
        this.template = template;
        this.point = point;
        this.format = format;
    }
    /**
     * @hidden
     */
    get formattedValue() {
        return this.format ? this.point.formatValue(this.format) : String(this.value);
    }
}
