// @progress/kendo-angular-charts/esm2022/events/series-click-event.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { BaseEvent } from './base-event';
/**
 * Represents the arguments for the `seriesClick` event of the Chart component ([see example](slug:events_chart)).
 */
export class SeriesClickEvent extends BaseEvent {
    /**
     * Specifies the data point category.
     */
    category;
    /**
     * Specifies the original data item.
     */
    dataItem;
    /**
     * Specifies the original user event that triggered the click action.
     */
    originalEvent;
    /**
     * Specifies the point value represented as a percentage value.
     * Available only for Donut, Pie, and 100% stacked charts.
     */
    percentage;
    /**
     * Specifies the clicked series point.
     */
    point;
    /**
     * Specifies the clicked point series options.
     */
    series;
    /**
     * Specifies the cumulative point value on the stack.
     * Available only for stackable series.
     */
    stackValue;
    /**
     * Specifies the data point value.
     */
    value;
    /**
     * @hidden
     */
    constructor(e, sender) {
        super(sender);
        this.category = e.category;
        this.dataItem = e.dataItem;
        this.originalEvent = e.originalEvent;
        this.percentage = e.percentage;
        this.point = e.point;
        this.series = e.series;
        this.stackValue = e.stackValue;
        this.value = e.value;
    }
}
