// @progress/kendo-angular-buttons/esm2022/splitbutton/splitbutton.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_SPLITBUTTON } from '../directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { KENDO_TOGGLEBUTTONTABSTOP, ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./splitbutton.component";
import * as i2 from "./localization/custom-messages.component";
import * as i3 from "@progress/kendo-angular-common";
import * as i4 from "../listbutton/button-item-template.directive";
// IMPORTANT: NgModule export kept for backwards compatibility.
/**
 * Represents the exported package module.
 *
 * Required for adding SplitButton features in NgModule-based Angular applications.
 *
 * The package exports:
 * - `SplitButtonComponent`&mdash;The SplitButton component class.
 * - `SplitButtonCustomMessagesComponent`&mdash;The SplitButton custom messages component class.
 */
export class SplitButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SplitButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: SplitButtonModule, imports: [i1.SplitButtonComponent, i2.SplitButtonCustomMessagesComponent, i3.ToggleButtonTabStopDirective, i4.ButtonItemTemplateDirective, i3.ToggleButtonTabStopDirective], exports: [i1.SplitButtonComponent, i2.SplitButtonCustomMessagesComponent, i3.ToggleButtonTabStopDirective, i4.ButtonItemTemplateDirective, i3.ToggleButtonTabStopDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SplitButtonModule, providers: [IconsService, PopupService, ResizeBatchService], imports: [i1.SplitButtonComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SplitButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [...KENDO_SPLITBUTTON, ...KENDO_TOGGLEBUTTONTABSTOP],
                    imports: [...KENDO_SPLITBUTTON, ...KENDO_TOGGLEBUTTONTABSTOP],
                    providers: [IconsService, PopupService, ResizeBatchService]
                }]
        }] });
