// @progress/kendo-angular-buttons/esm2022/buttons.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_BUTTONS } from './directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { KENDO_TOGGLEBUTTONTABSTOP, ResizeBatchService } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./button/button.component";
import * as i2 from "./buttongroup/buttongroup.component";
import * as i3 from "./dropdownbutton/dropdownbutton.component";
import * as i4 from "./listbutton/button-item-template.directive";
import * as i5 from "./chip/chip.component";
import * as i6 from "./chip/chip-list.component";
import * as i7 from "./floatingactionbutton/floatingactionbutton.component";
import * as i8 from "./floatingactionbutton/templates/dial-item-template.directive";
import * as i9 from "./floatingactionbutton/templates/fab-template.directive";
import * as i10 from "./splitbutton/splitbutton.component";
import * as i11 from "./splitbutton/localization/custom-messages.component";
import * as i12 from "@progress/kendo-angular-common";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the [NgModule](link:site.data.urls.angular['ngmodules'])
 * definition for the Buttons components.
 *
 * Required for adding all Button features in NgModule-based Angular applications.
 *
 * @example
 *
 * ```ts-no-run
 * import { ButtonsModule } from '@progress/kendo-angular-buttons';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, ButtonsModule], // import Buttons module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class ButtonsModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ButtonsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: ButtonsModule, imports: [i1.ButtonComponent, i1.ButtonComponent, i2.ButtonGroupComponent, i3.DropDownButtonComponent, i4.ButtonItemTemplateDirective, i5.ChipComponent, i5.ChipComponent, i6.ChipListComponent, i7.FloatingActionButtonComponent, i8.DialItemTemplateDirective, i9.FloatingActionButtonTemplateDirective, i10.SplitButtonComponent, i11.SplitButtonCustomMessagesComponent, i12.ToggleButtonTabStopDirective, i4.ButtonItemTemplateDirective, i12.ToggleButtonTabStopDirective], exports: [i1.ButtonComponent, i1.ButtonComponent, i2.ButtonGroupComponent, i3.DropDownButtonComponent, i4.ButtonItemTemplateDirective, i5.ChipComponent, i5.ChipComponent, i6.ChipListComponent, i7.FloatingActionButtonComponent, i8.DialItemTemplateDirective, i9.FloatingActionButtonTemplateDirective, i10.SplitButtonComponent, i11.SplitButtonCustomMessagesComponent, i12.ToggleButtonTabStopDirective, i4.ButtonItemTemplateDirective, i12.ToggleButtonTabStopDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ButtonsModule, providers: [IconsService, PopupService, ResizeBatchService], imports: [i1.ButtonComponent, i1.ButtonComponent, i3.DropDownButtonComponent, i5.ChipComponent, i5.ChipComponent, i7.FloatingActionButtonComponent, i10.SplitButtonComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ButtonsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_BUTTONS, ...KENDO_TOGGLEBUTTONTABSTOP],
                    exports: [...KENDO_BUTTONS, ...KENDO_TOGGLEBUTTONTABSTOP],
                    providers: [IconsService, PopupService, ResizeBatchService]
                }]
        }] });
