// @progress/kendo-angular-dropdowns/esm2022/comboboxes/combobox-column/column-cell-template.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Renders the column cell content of the MultiColumnComboBox component.
 *
 * To define a column cell template, nest an `<ng-template>` tag with the `kendoMultiColumnComboBoxColumnCellTemplate` directive inside the [`<kendo-combobox-column>`]({% slug api_dropdowns_comboboxcolumncomponent %}) tag.
 * [See example]({% slug templates_multicolumncombobox %}).
 *
 * The context variables are:
 * - `let-dataItem="dataItem"` (`any`)&mdash;The current data item. Also available as the implicit context variable.
 * - `let-column="column"` ([`ColumnComponent`]({% slug api_dropdowns_comboboxcolumncomponent %}))&mdash;The current column configuration object.
 *
 * @example
 * ```html
 * <kendo-multicolumncombobox ...>
 *   <kendo-combobox-column field="name" title="Name">
 *     <ng-template kendoMultiColumnComboBoxColumnCellTemplate let-dataItem="dataItem" let-column="column">
 *       {{ dataItem[column.field] }}
 *     </ng-template>
 *   </kendo-combobox-column>
 * <kendo-multicolumncombobox>
 * ```
 */
export class ColumnCellTemplateDirective {
    templateRef;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ColumnCellTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: ColumnCellTemplateDirective, isStandalone: true, selector: "[kendoMultiColumnComboBoxColumnCellTemplate]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ColumnCellTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoMultiColumnComboBoxColumnCellTemplate]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef }]; } });
