// @progress/kendo-angular-inputs/esm2022/checkbox.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_CHECKBOX } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./checkbox/checkbox.component";
import * as i2 from "./checkbox/checkbox.directive";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the CheckBox directive and CheckBoxComponent.
 *
 * Use this module to add CheckBox features to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { CheckBoxModule } from '@progress/kendo-angular-inputs';
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule, CheckBoxModule],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class CheckBoxModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CheckBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: CheckBoxModule, imports: [i1.CheckBoxComponent, i2.CheckBoxDirective], exports: [i1.CheckBoxComponent, i2.CheckBoxDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CheckBoxModule });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CheckBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_CHECKBOX],
                    exports: [...KENDO_CHECKBOX]
                }]
        }] });
