// @progress/kendo-angular-inputs/esm2022/numerictextbox.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_NUMERICTEXTBOX } from './directives';
import { IconsService } from '@progress/kendo-angular-icons';
import { KENDO_ADORNMENTS } from '@progress/kendo-angular-common';
import * as i0 from "@angular/core";
import * as i1 from "./numerictextbox/numerictextbox.component";
import * as i2 from "./numerictextbox/localization/custom-messages.component";
import * as i3 from "@progress/kendo-angular-common";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the NumericTextBox component.
 *
 * Use this module to add the NumericTextBox component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { NumericTextBoxModule } from '@progress/kendo-angular-inputs';
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule, NumericTextBoxModule],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class NumericTextBoxModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NumericTextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: NumericTextBoxModule, imports: [i1.NumericTextBoxComponent, i2.NumericTextBoxCustomMessagesComponent, i3.PrefixTemplateDirective, i3.SuffixTemplateDirective, i3.SeparatorComponent, i3.PrefixTemplateDirective, i3.SuffixTemplateDirective, i3.SeparatorComponent], exports: [i1.NumericTextBoxComponent, i2.NumericTextBoxCustomMessagesComponent, i3.PrefixTemplateDirective, i3.SuffixTemplateDirective, i3.SeparatorComponent, i3.PrefixTemplateDirective, i3.SuffixTemplateDirective, i3.SeparatorComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NumericTextBoxModule, providers: [IconsService], imports: [i1.NumericTextBoxComponent, i3.SeparatorComponent, i3.SeparatorComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NumericTextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_NUMERICTEXTBOX, ...KENDO_ADORNMENTS],
                    exports: [...KENDO_NUMERICTEXTBOX, ...KENDO_ADORNMENTS],
                    providers: [IconsService]
                }]
        }] });
