// @progress/kendo-angular-inputs/esm2022/otp.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { KENDO_OTPINPUT } from "./directives";
import { IconsService } from "@progress/kendo-angular-icons";
import * as i0 from "@angular/core";
import * as i1 from "./otpinput/otpinput.component";
import * as i2 from "./otpinput/localization/custom-messages.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [NgModule](link:site.data.urls.angular['ngmoduleapi']) for the OTP Input component.
 *
 * Use this module to add the OTP Input component to your NgModule-based Angular application.
 *
 * @example
 * ```typescript
 * import { OTPInputModule } from '@progress/kendo-angular-inputs';
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *   declarations: [AppComponent],
 *   imports: [BrowserModule, OTPInputModule],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class OTPInputModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: OTPInputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: OTPInputModule, imports: [i1.OTPInputComponent, i2.OTPInputCustomMessagesComponent], exports: [i1.OTPInputComponent, i2.OTPInputCustomMessagesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: OTPInputModule, providers: [IconsService], imports: [i1.OTPInputComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: OTPInputModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_OTPINPUT],
                    exports: [...KENDO_OTPINPUT],
                    providers: [IconsService]
                }]
        }] });
