// @progress/kendo-angular-inputs/esm2022/rating.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { KENDO_RATING } from './directives';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "./rating/rating.component";
import * as i2 from "./rating/directives/rating-item.directive";
import * as i3 from "./rating/directives/rating-hovered-item.directive";
import * as i4 from "./rating/directives/rating-selected-item.directive";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the Rating component.
 *
 * Use this module to add the Rating component to your NgModule-based Angular application.
 *
 * @example
 *
 * ```typescript
 * import { RatingModule } from '@progress/kendo-angular-inputs';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, RatingModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class RatingModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: RatingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: RatingModule, imports: [i1.RatingComponent, i2.RatingItemTemplateDirective, i3.RatingHoveredItemTemplateDirective, i4.RatingSelectedItemTemplateDirective], exports: [i1.RatingComponent, i2.RatingItemTemplateDirective, i3.RatingHoveredItemTemplateDirective, i4.RatingSelectedItemTemplateDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: RatingModule, providers: [IconsService], imports: [i1.RatingComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: RatingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_RATING],
                    exports: [...KENDO_RATING],
                    providers: [IconsService]
                }]
        }] });
