// @progress/kendo-angular-inputs/esm2022/textbox.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from "@angular/core";
import { KENDO_TEXTBOX } from "./directives";
import { IconsService } from "@progress/kendo-angular-icons";
import { KENDO_ADORNMENTS } from "@progress/kendo-angular-common";
import * as i0 from "@angular/core";
import * as i1 from "./textbox/textbox.directive";
import * as i2 from "./textbox/textbox.component";
import * as i3 from "./shared/input-separator.component";
import * as i4 from "./textbox/textbox-suffix.directive";
import * as i5 from "./textbox/textbox-prefix.directive";
import * as i6 from "./textbox/localization/custom-messages.component";
import * as i7 from "@progress/kendo-angular-common";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Defines the [`NgModule`](link:site.data.urls.angular['ngmoduleapi']) for the TextBox directive.
 *
 * Use this module to add the TextBox directive to your NgModule-based Angular application.
 *
 * @example
 *
 * ```typescript
 * import { TextBoxModule } from '@progress/kendo-angular-inputs';
 * import { NgModule } from '@angular/core';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, TextBoxModule],
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 * ```
 */
export class TextBoxModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextBoxModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: TextBoxModule, imports: [i1.TextBoxDirective, i2.TextBoxComponent, i3.InputSeparatorComponent, i4.TextBoxSuffixTemplateDirective, i5.TextBoxPrefixTemplateDirective, i6.TextBoxCustomMessagesComponent, i7.PrefixTemplateDirective, i7.SuffixTemplateDirective, i7.SeparatorComponent, i7.PrefixTemplateDirective, i7.SuffixTemplateDirective, i7.SeparatorComponent], exports: [i1.TextBoxDirective, i2.TextBoxComponent, i3.InputSeparatorComponent, i4.TextBoxSuffixTemplateDirective, i5.TextBoxPrefixTemplateDirective, i6.TextBoxCustomMessagesComponent, i7.PrefixTemplateDirective, i7.SuffixTemplateDirective, i7.SeparatorComponent, i7.PrefixTemplateDirective, i7.SuffixTemplateDirective, i7.SeparatorComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextBoxModule, providers: [IconsService], imports: [i2.TextBoxComponent, i7.SeparatorComponent, i7.SeparatorComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: TextBoxModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_TEXTBOX, ...KENDO_ADORNMENTS],
                    exports: [...KENDO_TEXTBOX, ...KENDO_ADORNMENTS],
                    providers: [IconsService]
                }]
        }] });
