// @progress/kendo-angular-charts/esm2022/chart.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { PopupService } from '@progress/kendo-angular-popup';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { ThemeService } from './common/theme.service';
import { KENDO_CHART } from './directives';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "./chart.component";
import * as i2 from "./chart/donut-center-template.directive";
import * as i3 from "./chart/no-data-template.directive";
import * as i4 from "./chart/x-axis.component";
import * as i5 from "./chart/x-axis-item/crosshair.component";
import * as i6 from "./chart/x-axis-item/crosshair.tooltip.component";
import * as i7 from "./chart/x-axis-item.component";
import * as i8 from "./chart/x-axis-item/labels.component";
import * as i9 from "./chart/x-axis-item/notes.component";
import * as i10 from "./chart/x-axis-item/notes.icon.component";
import * as i11 from "./chart/x-axis-item/notes.label.component";
import * as i12 from "./chart/x-axis-item/title.component";
import * as i13 from "./chart/y-axis.component";
import * as i14 from "./chart/y-axis-item/crosshair.component";
import * as i15 from "./chart/y-axis-item/crosshair.tooltip.component";
import * as i16 from "./chart/y-axis-item.component";
import * as i17 from "./chart/y-axis-item/labels.component";
import * as i18 from "./chart/y-axis-item/notes.component";
import * as i19 from "./chart/y-axis-item/notes.icon.component";
import * as i20 from "./chart/y-axis-item/notes.label.component";
import * as i21 from "./chart/y-axis-item/title.component";
import * as i22 from "./chart/zoomable.component";
import * as i23 from "./chart/tooltip/tooltip-popup.component";
import * as i24 from "./chart/tooltip/series-tooltip-template.directive";
import * as i25 from "./chart/tooltip/shared-tooltip-template.directive";
import * as i26 from "./chart/tooltip/crosshair-tooltips-container.component";
import * as i27 from "./chart/tooltip/crosshair-tooltip.component";
import * as i28 from "./chart/axis-defaults.component";
import * as i29 from "./chart/axis-defaults/crosshair.component";
import * as i30 from "./chart/axis-defaults/crosshair.tooltip.component";
import * as i31 from "./chart/axis-defaults/labels.component";
import * as i32 from "./chart/axis-defaults/title.component";
import * as i33 from "./chart/category-axis.component";
import * as i34 from "./chart/category-axis-item/crosshair.component";
import * as i35 from "./chart/category-axis-item/crosshair.tooltip.component";
import * as i36 from "./chart/category-axis-item.component";
import * as i37 from "./chart/category-axis-item/labels.component";
import * as i38 from "./chart/category-axis-item/range-labels.component";
import * as i39 from "./chart/category-axis-item/notes.component";
import * as i40 from "./chart/category-axis-item/notes.icon.component";
import * as i41 from "./chart/category-axis-item/notes.label.component";
import * as i42 from "./chart/category-axis-item/select.component";
import * as i43 from "./chart/category-axis-item/title.component";
import * as i44 from "./chart/chart-area.component";
import * as i45 from "./chart-breadcrumb.component";
import * as i46 from "./chart/localization/custom-messages.component";
import * as i47 from "./chart/localization/localized-messages.directive";
import * as i48 from "./chart/legend.component";
import * as i49 from "./chart/legend/inactive-items.component";
import * as i50 from "./chart/legend/item.component";
import * as i51 from "./chart/pane.component";
import * as i52 from "./chart/pane-defaults.component";
import * as i53 from "./chart/pane-defaults/title.component";
import * as i54 from "./chart/panes.component";
import * as i55 from "./chart/pane/title.component";
import * as i56 from "./chart/plot-area.component";
import * as i57 from "./chart/series.component";
import * as i58 from "./chart/series-defaults.component";
import * as i59 from "./chart/series-defaults/labels.component";
import * as i60 from "./chart/series-defaults/labels.from.component";
import * as i61 from "./chart/series-defaults/labels.to.component";
import * as i62 from "./chart/series-defaults/notes.component";
import * as i63 from "./chart/series-defaults/notes.icon.component";
import * as i64 from "./chart/series-defaults/notes.label.component";
import * as i65 from "./chart/series-defaults/tooltip.component";
import * as i66 from "./chart/series-drilldown-template.directive";
import * as i67 from "./chart/series-item/error-bars.component";
import * as i68 from "./chart/series-item/extremes.component";
import * as i69 from "./chart/series-item/highlight.component";
import * as i70 from "./chart/series-item.component";
import * as i71 from "./chart/series-item/labels.component";
import * as i72 from "./chart/series-item/labels.from.component";
import * as i73 from "./chart/series-item/labels.to.component";
import * as i74 from "./chart/series-item/markers.component";
import * as i75 from "./chart/series-item/notes.component";
import * as i76 from "./chart/series-item/notes.icon.component";
import * as i77 from "./chart/series-item/notes.label.component";
import * as i78 from "./chart/series-item/outliers.component";
import * as i79 from "./chart/series-item/tooltip.component";
import * as i80 from "./chart/series-item/trendline.component";
import * as i81 from "./chart/series-item/trendline.forecast.component";
import * as i82 from "./chart/subtitle.component";
import * as i83 from "./chart/title.component";
import * as i84 from "./chart/tooltip.component";
import * as i85 from "./chart/value-axis.component";
import * as i86 from "./chart/value-axis-item/crosshair.component";
import * as i87 from "./chart/value-axis-item/crosshair.tooltip.component";
import * as i88 from "./chart/value-axis-item.component";
import * as i89 from "./chart/value-axis-item/labels.component";
import * as i90 from "./chart/value-axis-item/notes.component";
import * as i91 from "./chart/value-axis-item/notes.icon.component";
import * as i92 from "./chart/value-axis-item/notes.label.component";
import * as i93 from "./chart/value-axis-item/title.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the Kendo UI for Angular Chart module that includes the Chart component and directives.
 *
 * Import the `ChartModule` into your application
 * [root module](link:site.data.urls.angular['ngmodules']#angular-modularity) or any other sub-module
 * that uses the Chart component.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { ChartModule } from '@progress/kendo-angular-charts';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     bootstrap:    [AppComponent],
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, ChartModule]
 * })
 * export class AppModule {
 * }
 * ```
 */
export class ChartModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: ChartModule, imports: [i1.ChartComponent, i2.DonutCenterTemplateDirective, i3.NoDataTemplateDirective, i4.XAxisComponent, i5.XAxisCrosshairComponent, i6.XAxisCrosshairTooltipComponent, i7.XAxisItemComponent, i8.XAxisLabelsComponent, i9.XAxisNotesComponent, i10.XAxisNotesIconComponent, i11.XAxisNotesLabelComponent, i12.XAxisTitleComponent, i13.YAxisComponent, i14.YAxisCrosshairComponent, i15.YAxisCrosshairTooltipComponent, i16.YAxisItemComponent, i17.YAxisLabelsComponent, i18.YAxisNotesComponent, i19.YAxisNotesIconComponent, i20.YAxisNotesLabelComponent, i21.YAxisTitleComponent, i22.ZoomableComponent, i23.TooltipPopupComponent, i24.SeriesTooltipTemplateDirective, i25.SharedTooltipTemplateDirective, i26.CrosshairTooltipsContainerComponent, i27.CrosshairTooltipComponent, i28.AxisDefaultsComponent, i29.AxisDefaultsCrosshairComponent, i30.AxisDefaultsCrosshairTooltipComponent, i31.AxisDefaultsLabelsComponent, i32.AxisDefaultsTitleComponent, i33.CategoryAxisComponent, i34.CategoryAxisCrosshairComponent, i35.CategoryAxisCrosshairTooltipComponent, i36.CategoryAxisItemComponent, i37.CategoryAxisLabelsComponent, i38.CategoryAxisRangeLabelsComponent, i39.CategoryAxisNotesComponent, i40.CategoryAxisNotesIconComponent, i41.CategoryAxisNotesLabelComponent, i42.CategoryAxisSelectComponent, i43.CategoryAxisTitleComponent, i44.ChartAreaComponent, i45.ChartBreadcrumbComponent, i46.ChartCustomMessagesComponent, i47.LocalizedChartMessagesDirective, i48.LegendComponent, i49.LegendInactiveItemsComponent, i50.LegendItemComponent, i51.PaneComponent, i52.PaneDefaultsComponent, i53.PaneDefaultsTitleComponent, i54.PanesComponent, i55.PanesTitleComponent, i56.PlotAreaComponent, i57.SeriesComponent, i58.SeriesDefaultsComponent, i59.SeriesDefaultsLabelsComponent, i60.SeriesDefaultsLabelsFromComponent, i61.SeriesDefaultsLabelsToComponent, i62.SeriesDefaultsNotesComponent, i63.SeriesDefaultsNotesIconComponent, i64.SeriesDefaultsNotesLabelComponent, i65.SeriesDefaultsTooltipComponent, i66.SeriesDrilldownTemplateDirective, i67.SeriesErrorBarsComponent, i68.SeriesExtremesComponent, i69.SeriesHighlightComponent, i70.SeriesItemComponent, i71.SeriesLabelsComponent, i72.SeriesLabelsFromComponent, i73.SeriesLabelsToComponent, i74.SeriesMarkersComponent, i75.SeriesNotesComponent, i76.SeriesNotesIconComponent, i77.SeriesNotesLabelComponent, i78.SeriesOutliersComponent, i79.SeriesTooltipComponent, i80.SeriesTrendlineComponent, i81.SeriesTrendlineForecastComponent, i82.SubtitleComponent, i83.TitleComponent, i84.TooltipComponent, i85.ValueAxisComponent, i86.ValueAxisCrosshairComponent, i87.ValueAxisCrosshairTooltipComponent, i88.ValueAxisItemComponent, i89.ValueAxisLabelsComponent, i90.ValueAxisNotesComponent, i91.ValueAxisNotesIconComponent, i92.ValueAxisNotesLabelComponent, i93.ValueAxisTitleComponent], exports: [i1.ChartComponent, i2.DonutCenterTemplateDirective, i3.NoDataTemplateDirective, i4.XAxisComponent, i5.XAxisCrosshairComponent, i6.XAxisCrosshairTooltipComponent, i7.XAxisItemComponent, i8.XAxisLabelsComponent, i9.XAxisNotesComponent, i10.XAxisNotesIconComponent, i11.XAxisNotesLabelComponent, i12.XAxisTitleComponent, i13.YAxisComponent, i14.YAxisCrosshairComponent, i15.YAxisCrosshairTooltipComponent, i16.YAxisItemComponent, i17.YAxisLabelsComponent, i18.YAxisNotesComponent, i19.YAxisNotesIconComponent, i20.YAxisNotesLabelComponent, i21.YAxisTitleComponent, i22.ZoomableComponent, i23.TooltipPopupComponent, i24.SeriesTooltipTemplateDirective, i25.SharedTooltipTemplateDirective, i26.CrosshairTooltipsContainerComponent, i27.CrosshairTooltipComponent, i28.AxisDefaultsComponent, i29.AxisDefaultsCrosshairComponent, i30.AxisDefaultsCrosshairTooltipComponent, i31.AxisDefaultsLabelsComponent, i32.AxisDefaultsTitleComponent, i33.CategoryAxisComponent, i34.CategoryAxisCrosshairComponent, i35.CategoryAxisCrosshairTooltipComponent, i36.CategoryAxisItemComponent, i37.CategoryAxisLabelsComponent, i38.CategoryAxisRangeLabelsComponent, i39.CategoryAxisNotesComponent, i40.CategoryAxisNotesIconComponent, i41.CategoryAxisNotesLabelComponent, i42.CategoryAxisSelectComponent, i43.CategoryAxisTitleComponent, i44.ChartAreaComponent, i45.ChartBreadcrumbComponent, i46.ChartCustomMessagesComponent, i47.LocalizedChartMessagesDirective, i48.LegendComponent, i49.LegendInactiveItemsComponent, i50.LegendItemComponent, i51.PaneComponent, i52.PaneDefaultsComponent, i53.PaneDefaultsTitleComponent, i54.PanesComponent, i55.PanesTitleComponent, i56.PlotAreaComponent, i57.SeriesComponent, i58.SeriesDefaultsComponent, i59.SeriesDefaultsLabelsComponent, i60.SeriesDefaultsLabelsFromComponent, i61.SeriesDefaultsLabelsToComponent, i62.SeriesDefaultsNotesComponent, i63.SeriesDefaultsNotesIconComponent, i64.SeriesDefaultsNotesLabelComponent, i65.SeriesDefaultsTooltipComponent, i66.SeriesDrilldownTemplateDirective, i67.SeriesErrorBarsComponent, i68.SeriesExtremesComponent, i69.SeriesHighlightComponent, i70.SeriesItemComponent, i71.SeriesLabelsComponent, i72.SeriesLabelsFromComponent, i73.SeriesLabelsToComponent, i74.SeriesMarkersComponent, i75.SeriesNotesComponent, i76.SeriesNotesIconComponent, i77.SeriesNotesLabelComponent, i78.SeriesOutliersComponent, i79.SeriesTooltipComponent, i80.SeriesTrendlineComponent, i81.SeriesTrendlineForecastComponent, i82.SubtitleComponent, i83.TitleComponent, i84.TooltipComponent, i85.ValueAxisComponent, i86.ValueAxisCrosshairComponent, i87.ValueAxisCrosshairTooltipComponent, i88.ValueAxisItemComponent, i89.ValueAxisLabelsComponent, i90.ValueAxisNotesComponent, i91.ValueAxisNotesIconComponent, i92.ValueAxisNotesLabelComponent, i93.ValueAxisTitleComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartModule, providers: [IconsService, ResizeBatchService, PopupService, ThemeService], imports: [i1.ChartComponent, i45.ChartBreadcrumbComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: ChartModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_CHART],
                    exports: [...KENDO_CHART],
                    providers: [IconsService, ResizeBatchService, PopupService, ThemeService]
                }]
        }] });
