// @progress/kendo-angular-charts/esm2022/sankey.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { IconsService } from '@progress/kendo-angular-icons';
import { PopupService } from '@progress/kendo-angular-popup';
import { ThemeService } from './common/theme.service';
import { KENDO_SANKEY } from './directives';
import * as i0 from "@angular/core";
import * as i1 from "./sankey.component";
import * as i2 from "./sankey/tooltip/tooltip-popup.component";
import * as i3 from "./sankey/tooltip/link-tooltip-template.directive";
import * as i4 from "./sankey/tooltip/node-tooltip-template.directive";
import * as i5 from "./sankey/localization/custom-messages.component";
import * as i6 from "./sankey/localization/localized-messages.directive";
import * as i7 from "./sankey/title.component";
import * as i8 from "./sankey/legend.component";
import * as i9 from "./sankey/tooltip.component";
import * as i10 from "./sankey/links.component";
import * as i11 from "./sankey/nodes.component";
import * as i12 from "./sankey/labels.component";
import * as i13 from "./sankey/data-binding/flat-binding.directive";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the Kendo UI for Angular Sankey module that includes the Sankey component and directives.
 *
 * Import the `SankeyModule` into your application
 * [root module](link:site.data.urls.angular['ngmodules']#angular-modularity) or any other sub-module
 * that uses the Sankey component.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { SankeyModule } from '@progress/kendo-angular-charts';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     bootstrap:    [AppComponent],
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, SankeyModule]
 * })
 * export class AppModule {
 * }
 * ```
 */
export class SankeyModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: SankeyModule, imports: [i1.SankeyComponent, i2.SankeyTooltipPopupComponent, i3.SankeyLinkTooltipTemplateDirective, i4.SankeyNodeTooltipTemplateDirective, i5.SankeyCustomMessagesComponent, i6.LocalizedMessagesDirective, i7.SankeyTitleComponent, i8.SankeyLegendComponent, i9.SankeyTooltipComponent, i10.SankeyLinksComponent, i11.SankeyNodesComponent, i12.SankeyLabelsComponent, i13.SankeyFlatBindingDirective], exports: [i1.SankeyComponent, i2.SankeyTooltipPopupComponent, i3.SankeyLinkTooltipTemplateDirective, i4.SankeyNodeTooltipTemplateDirective, i5.SankeyCustomMessagesComponent, i6.LocalizedMessagesDirective, i7.SankeyTitleComponent, i8.SankeyLegendComponent, i9.SankeyTooltipComponent, i10.SankeyLinksComponent, i11.SankeyNodesComponent, i12.SankeyLabelsComponent, i13.SankeyFlatBindingDirective] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyModule, providers: [IconsService, ResizeBatchService, PopupService, ThemeService], imports: [i1.SankeyComponent, i2.SankeyTooltipPopupComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_SANKEY],
                    exports: [...KENDO_SANKEY],
                    providers: [IconsService, ResizeBatchService, PopupService, ThemeService]
                }]
        }] });
