// @progress/kendo-angular-charts/esm2022/sparkline.module.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { NgModule } from '@angular/core';
import { PopupService } from '@progress/kendo-angular-popup';
import { ResizeBatchService } from '@progress/kendo-angular-common';
import { KENDO_CHART, KENDO_SPARKLINE } from './directives';
import { ThemeService } from './common/theme.service';
import { IconsService } from '@progress/kendo-angular-icons';
import * as i0 from "@angular/core";
import * as i1 from "./sparkline.component";
import * as i2 from "./chart.component";
import * as i3 from "./chart/donut-center-template.directive";
import * as i4 from "./chart/no-data-template.directive";
import * as i5 from "./chart/x-axis.component";
import * as i6 from "./chart/x-axis-item/crosshair.component";
import * as i7 from "./chart/x-axis-item/crosshair.tooltip.component";
import * as i8 from "./chart/x-axis-item.component";
import * as i9 from "./chart/x-axis-item/labels.component";
import * as i10 from "./chart/x-axis-item/notes.component";
import * as i11 from "./chart/x-axis-item/notes.icon.component";
import * as i12 from "./chart/x-axis-item/notes.label.component";
import * as i13 from "./chart/x-axis-item/title.component";
import * as i14 from "./chart/y-axis.component";
import * as i15 from "./chart/y-axis-item/crosshair.component";
import * as i16 from "./chart/y-axis-item/crosshair.tooltip.component";
import * as i17 from "./chart/y-axis-item.component";
import * as i18 from "./chart/y-axis-item/labels.component";
import * as i19 from "./chart/y-axis-item/notes.component";
import * as i20 from "./chart/y-axis-item/notes.icon.component";
import * as i21 from "./chart/y-axis-item/notes.label.component";
import * as i22 from "./chart/y-axis-item/title.component";
import * as i23 from "./chart/zoomable.component";
import * as i24 from "./chart/tooltip/tooltip-popup.component";
import * as i25 from "./chart/tooltip/series-tooltip-template.directive";
import * as i26 from "./chart/tooltip/shared-tooltip-template.directive";
import * as i27 from "./chart/tooltip/crosshair-tooltips-container.component";
import * as i28 from "./chart/tooltip/crosshair-tooltip.component";
import * as i29 from "./chart/axis-defaults.component";
import * as i30 from "./chart/axis-defaults/crosshair.component";
import * as i31 from "./chart/axis-defaults/crosshair.tooltip.component";
import * as i32 from "./chart/axis-defaults/labels.component";
import * as i33 from "./chart/axis-defaults/title.component";
import * as i34 from "./chart/category-axis.component";
import * as i35 from "./chart/category-axis-item/crosshair.component";
import * as i36 from "./chart/category-axis-item/crosshair.tooltip.component";
import * as i37 from "./chart/category-axis-item.component";
import * as i38 from "./chart/category-axis-item/labels.component";
import * as i39 from "./chart/category-axis-item/range-labels.component";
import * as i40 from "./chart/category-axis-item/notes.component";
import * as i41 from "./chart/category-axis-item/notes.icon.component";
import * as i42 from "./chart/category-axis-item/notes.label.component";
import * as i43 from "./chart/category-axis-item/select.component";
import * as i44 from "./chart/category-axis-item/title.component";
import * as i45 from "./chart/chart-area.component";
import * as i46 from "./chart-breadcrumb.component";
import * as i47 from "./chart/localization/custom-messages.component";
import * as i48 from "./chart/localization/localized-messages.directive";
import * as i49 from "./chart/legend.component";
import * as i50 from "./chart/legend/inactive-items.component";
import * as i51 from "./chart/legend/item.component";
import * as i52 from "./chart/pane.component";
import * as i53 from "./chart/pane-defaults.component";
import * as i54 from "./chart/pane-defaults/title.component";
import * as i55 from "./chart/panes.component";
import * as i56 from "./chart/pane/title.component";
import * as i57 from "./chart/plot-area.component";
import * as i58 from "./chart/series.component";
import * as i59 from "./chart/series-defaults.component";
import * as i60 from "./chart/series-defaults/labels.component";
import * as i61 from "./chart/series-defaults/labels.from.component";
import * as i62 from "./chart/series-defaults/labels.to.component";
import * as i63 from "./chart/series-defaults/notes.component";
import * as i64 from "./chart/series-defaults/notes.icon.component";
import * as i65 from "./chart/series-defaults/notes.label.component";
import * as i66 from "./chart/series-defaults/tooltip.component";
import * as i67 from "./chart/series-drilldown-template.directive";
import * as i68 from "./chart/series-item/error-bars.component";
import * as i69 from "./chart/series-item/extremes.component";
import * as i70 from "./chart/series-item/highlight.component";
import * as i71 from "./chart/series-item.component";
import * as i72 from "./chart/series-item/labels.component";
import * as i73 from "./chart/series-item/labels.from.component";
import * as i74 from "./chart/series-item/labels.to.component";
import * as i75 from "./chart/series-item/markers.component";
import * as i76 from "./chart/series-item/notes.component";
import * as i77 from "./chart/series-item/notes.icon.component";
import * as i78 from "./chart/series-item/notes.label.component";
import * as i79 from "./chart/series-item/outliers.component";
import * as i80 from "./chart/series-item/tooltip.component";
import * as i81 from "./chart/series-item/trendline.component";
import * as i82 from "./chart/series-item/trendline.forecast.component";
import * as i83 from "./chart/subtitle.component";
import * as i84 from "./chart/title.component";
import * as i85 from "./chart/tooltip.component";
import * as i86 from "./chart/value-axis.component";
import * as i87 from "./chart/value-axis-item/crosshair.component";
import * as i88 from "./chart/value-axis-item/crosshair.tooltip.component";
import * as i89 from "./chart/value-axis-item.component";
import * as i90 from "./chart/value-axis-item/labels.component";
import * as i91 from "./chart/value-axis-item/notes.component";
import * as i92 from "./chart/value-axis-item/notes.icon.component";
import * as i93 from "./chart/value-axis-item/notes.label.component";
import * as i94 from "./chart/value-axis-item/title.component";
//IMPORTANT: NgModule export kept for backwards compatibility
/**
 * Represents the Kendo UI for Angular Sparkline module that includes the Sparkline component and directives.
 *
 * Import the `SparklineModule` into your application
 * [root module](link:site.data.urls.angular['ngmodules']#angular-modularity) or any other sub-module
 * that uses the Sparkline component.
 *
 * @example
 * ```ts
 * import { NgModule } from '@angular/core';
 * import { BrowserModule } from '@angular/platform-browser';
 * import { SparklineModule } from '@progress/kendo-angular-charts';
 * import { AppComponent } from './app.component';
 *
 * @NgModule({
 *     bootstrap:    [AppComponent],
 *     declarations: [AppComponent],
 *     imports:      [BrowserModule, SparklineModule]
 * })
 * export class AppModule {
 * }
 * ```
 */
export class SparklineModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SparklineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.2.12", ngImport: i0, type: SparklineModule, imports: [i1.SparklineComponent, i2.ChartComponent, i3.DonutCenterTemplateDirective, i4.NoDataTemplateDirective, i5.XAxisComponent, i6.XAxisCrosshairComponent, i7.XAxisCrosshairTooltipComponent, i8.XAxisItemComponent, i9.XAxisLabelsComponent, i10.XAxisNotesComponent, i11.XAxisNotesIconComponent, i12.XAxisNotesLabelComponent, i13.XAxisTitleComponent, i14.YAxisComponent, i15.YAxisCrosshairComponent, i16.YAxisCrosshairTooltipComponent, i17.YAxisItemComponent, i18.YAxisLabelsComponent, i19.YAxisNotesComponent, i20.YAxisNotesIconComponent, i21.YAxisNotesLabelComponent, i22.YAxisTitleComponent, i23.ZoomableComponent, i24.TooltipPopupComponent, i25.SeriesTooltipTemplateDirective, i26.SharedTooltipTemplateDirective, i27.CrosshairTooltipsContainerComponent, i28.CrosshairTooltipComponent, i29.AxisDefaultsComponent, i30.AxisDefaultsCrosshairComponent, i31.AxisDefaultsCrosshairTooltipComponent, i32.AxisDefaultsLabelsComponent, i33.AxisDefaultsTitleComponent, i34.CategoryAxisComponent, i35.CategoryAxisCrosshairComponent, i36.CategoryAxisCrosshairTooltipComponent, i37.CategoryAxisItemComponent, i38.CategoryAxisLabelsComponent, i39.CategoryAxisRangeLabelsComponent, i40.CategoryAxisNotesComponent, i41.CategoryAxisNotesIconComponent, i42.CategoryAxisNotesLabelComponent, i43.CategoryAxisSelectComponent, i44.CategoryAxisTitleComponent, i45.ChartAreaComponent, i46.ChartBreadcrumbComponent, i47.ChartCustomMessagesComponent, i48.LocalizedChartMessagesDirective, i49.LegendComponent, i50.LegendInactiveItemsComponent, i51.LegendItemComponent, i52.PaneComponent, i53.PaneDefaultsComponent, i54.PaneDefaultsTitleComponent, i55.PanesComponent, i56.PanesTitleComponent, i57.PlotAreaComponent, i58.SeriesComponent, i59.SeriesDefaultsComponent, i60.SeriesDefaultsLabelsComponent, i61.SeriesDefaultsLabelsFromComponent, i62.SeriesDefaultsLabelsToComponent, i63.SeriesDefaultsNotesComponent, i64.SeriesDefaultsNotesIconComponent, i65.SeriesDefaultsNotesLabelComponent, i66.SeriesDefaultsTooltipComponent, i67.SeriesDrilldownTemplateDirective, i68.SeriesErrorBarsComponent, i69.SeriesExtremesComponent, i70.SeriesHighlightComponent, i71.SeriesItemComponent, i72.SeriesLabelsComponent, i73.SeriesLabelsFromComponent, i74.SeriesLabelsToComponent, i75.SeriesMarkersComponent, i76.SeriesNotesComponent, i77.SeriesNotesIconComponent, i78.SeriesNotesLabelComponent, i79.SeriesOutliersComponent, i80.SeriesTooltipComponent, i81.SeriesTrendlineComponent, i82.SeriesTrendlineForecastComponent, i83.SubtitleComponent, i84.TitleComponent, i85.TooltipComponent, i86.ValueAxisComponent, i87.ValueAxisCrosshairComponent, i88.ValueAxisCrosshairTooltipComponent, i89.ValueAxisItemComponent, i90.ValueAxisLabelsComponent, i91.ValueAxisNotesComponent, i92.ValueAxisNotesIconComponent, i93.ValueAxisNotesLabelComponent, i94.ValueAxisTitleComponent], exports: [i1.SparklineComponent, i2.ChartComponent, i3.DonutCenterTemplateDirective, i4.NoDataTemplateDirective, i5.XAxisComponent, i6.XAxisCrosshairComponent, i7.XAxisCrosshairTooltipComponent, i8.XAxisItemComponent, i9.XAxisLabelsComponent, i10.XAxisNotesComponent, i11.XAxisNotesIconComponent, i12.XAxisNotesLabelComponent, i13.XAxisTitleComponent, i14.YAxisComponent, i15.YAxisCrosshairComponent, i16.YAxisCrosshairTooltipComponent, i17.YAxisItemComponent, i18.YAxisLabelsComponent, i19.YAxisNotesComponent, i20.YAxisNotesIconComponent, i21.YAxisNotesLabelComponent, i22.YAxisTitleComponent, i23.ZoomableComponent, i24.TooltipPopupComponent, i25.SeriesTooltipTemplateDirective, i26.SharedTooltipTemplateDirective, i27.CrosshairTooltipsContainerComponent, i28.CrosshairTooltipComponent, i29.AxisDefaultsComponent, i30.AxisDefaultsCrosshairComponent, i31.AxisDefaultsCrosshairTooltipComponent, i32.AxisDefaultsLabelsComponent, i33.AxisDefaultsTitleComponent, i34.CategoryAxisComponent, i35.CategoryAxisCrosshairComponent, i36.CategoryAxisCrosshairTooltipComponent, i37.CategoryAxisItemComponent, i38.CategoryAxisLabelsComponent, i39.CategoryAxisRangeLabelsComponent, i40.CategoryAxisNotesComponent, i41.CategoryAxisNotesIconComponent, i42.CategoryAxisNotesLabelComponent, i43.CategoryAxisSelectComponent, i44.CategoryAxisTitleComponent, i45.ChartAreaComponent, i46.ChartBreadcrumbComponent, i47.ChartCustomMessagesComponent, i48.LocalizedChartMessagesDirective, i49.LegendComponent, i50.LegendInactiveItemsComponent, i51.LegendItemComponent, i52.PaneComponent, i53.PaneDefaultsComponent, i54.PaneDefaultsTitleComponent, i55.PanesComponent, i56.PanesTitleComponent, i57.PlotAreaComponent, i58.SeriesComponent, i59.SeriesDefaultsComponent, i60.SeriesDefaultsLabelsComponent, i61.SeriesDefaultsLabelsFromComponent, i62.SeriesDefaultsLabelsToComponent, i63.SeriesDefaultsNotesComponent, i64.SeriesDefaultsNotesIconComponent, i65.SeriesDefaultsNotesLabelComponent, i66.SeriesDefaultsTooltipComponent, i67.SeriesDrilldownTemplateDirective, i68.SeriesErrorBarsComponent, i69.SeriesExtremesComponent, i70.SeriesHighlightComponent, i71.SeriesItemComponent, i72.SeriesLabelsComponent, i73.SeriesLabelsFromComponent, i74.SeriesLabelsToComponent, i75.SeriesMarkersComponent, i76.SeriesNotesComponent, i77.SeriesNotesIconComponent, i78.SeriesNotesLabelComponent, i79.SeriesOutliersComponent, i80.SeriesTooltipComponent, i81.SeriesTrendlineComponent, i82.SeriesTrendlineForecastComponent, i83.SubtitleComponent, i84.TitleComponent, i85.TooltipComponent, i86.ValueAxisComponent, i87.ValueAxisCrosshairComponent, i88.ValueAxisCrosshairTooltipComponent, i89.ValueAxisItemComponent, i90.ValueAxisLabelsComponent, i91.ValueAxisNotesComponent, i92.ValueAxisNotesIconComponent, i93.ValueAxisNotesLabelComponent, i94.ValueAxisTitleComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SparklineModule, providers: [ThemeService, IconsService, PopupService, ResizeBatchService], imports: [KENDO_SPARKLINE, i2.ChartComponent, i46.ChartBreadcrumbComponent] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SparklineModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...KENDO_SPARKLINE, ...KENDO_CHART],
                    exports: [...KENDO_SPARKLINE, ...KENDO_CHART],
                    providers: [ThemeService, IconsService, PopupService, ResizeBatchService]
                }]
        }] });
