// @progress/kendo-angular-charts/esm2022/chart/x-axis-item/notes.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../../common/configuration.service';
import { SettingsComponent } from '../../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
/**
 * Represents the configuration of the X-axis notes.
 * For an example on the basic usage of the `XAxisNotesComponent`,
 * refer to the [demo on the `XAxisComponent`]({% slug api_charts_xaxiscomponent %})
 * or to the documentation about the
 * [axis notes]({% slug notes_chart_charts %}#toc-axis-notes).
 *
 * @example
 * ```html
 * <kendo-chart>
 *  <kendo-chart-x-axis>
 *   <kendo-chart-x-axis-item>
 *    <kendo-chart-x-axis-item-notes>
 *    </kendo-chart-x-axis-item-notes>
 *  </kendo-chart-x-axis-item>
 *  </kendo-chart-x-axis>
 * </kendo-chart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link XAxisNotesIconComponent} and {@link XAxisNotesLabelComponent}.
 */
export class XAxisNotesComponent extends SettingsComponent {
    configurationService;
    /**
     * Specifies the data for the notes.
     */
    data;
    /**
     * Specifies the line configuration of the notes.
     */
    line;
    /**
     * Specifies the position of the notes.
     */
    position;
    /**
     * Specifies a function that returns a custom visual for the notes.
     */
    visual;
    // These options are also available as child components
    /**
     * Specifies the icon configuration of the notes.
     */
    icon;
    /**
     * Specifies the label configuration of the notes.
     */
    label;
    constructor(configurationService) {
        super('notes', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: XAxisNotesComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: XAxisNotesComponent, isStandalone: true, selector: "kendo-chart-x-axis-item-notes", inputs: { data: "data", line: "line", position: "position", visual: "visual", icon: "icon", label: "label" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: XAxisNotesComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-chart-x-axis-item-notes',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { data: [{
                type: Input
            }], line: [{
                type: Input
            }], position: [{
                type: Input
            }], visual: [{
                type: Input
            }], icon: [{
                type: Input
            }], label: [{
                type: Input
            }] } });
