// @progress/kendo-angular-charts/esm2022/stock-chart/navigator/series-item.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Component, ChangeDetectionStrategy, Input } from '@angular/core';
import { CollectionService } from '../../common/collection.service';
import { ConfigurationService } from '../../common/configuration.service';
import { SeriesItemComponent } from '../../chart/series-item.component';
import * as i0 from "@angular/core";
import * as i1 from "../../common/configuration.service";
import * as i2 from "../../common/collection.service";
/**
 * Represents a series item in the StockChart navigator
 * ([see example]({% slug overview_stockchart_charts %})).
 *
 * @example
 * ```html
 * <kendo-stockchart>
 *  <kendo-chart-navigator>
 *   <kendo-chart-navigator-series>
 *    <kendo-chart-navigator-series-item type="area" [data]="data" field="value" categoryField="date">
 *    </kendo-chart-navigator-series-item>
 *   </kendo-chart-navigator-series>
 *  </kendo-chart-navigator>
 * <kendo-stockchart>
 * ```
 *
 * @remarks
 * Supported children components are: {@link NavigatorSeriesHighlightComponent}, {@link NavigatorSeriesErrorBarsComponent}, {@link NavigatorSeriesExtremesComponent},
 * {@link NavigatorSeriesLabelsComponent}, {@link NavigatorSeriesMarkersComponent}, {@link NavigatorSeriesNotesComponent}, and {@link NavigatorSeriesTooltipComponent}.
 */
export class NavigatorSeriesItemComponent extends SeriesItemComponent {
    configurationService;
    collectionService;
    /**
     * @hidden
     */
    drilldownField;
    constructor(configurationService, collectionService) {
        super(configurationService, collectionService);
        this.configurationService = configurationService;
        this.collectionService = collectionService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorSeriesItemComponent, deps: [{ token: i1.ConfigurationService }, { token: i2.CollectionService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: NavigatorSeriesItemComponent, isStandalone: true, selector: "kendo-chart-navigator-series-item", inputs: { drilldownField: "drilldownField" }, providers: [ConfigurationService], usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: NavigatorSeriesItemComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [ConfigurationService],
                    selector: 'kendo-chart-navigator-series-item',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }, { type: i2.CollectionService }]; }, propDecorators: { drilldownField: [{
                type: Input
            }] } });
