// @progress/kendo-angular-common/esm2022/adornments/suffix.directive.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, Input, Optional, TemplateRef } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Represents the directive for suffix adornments in the Inputs and DropDowns components.
 *
 * Use the `kendoSuffixTemplate` directive to add custom content to the suffix container.
 *
 * * See [Inputs Suffix Adornments](slug:adornments_textbox#toc-suffix-adornments)
 * * See [DropDowns Suffix Adornments](slug:adornments_multiselect#toc-suffix-adornments)
 *
 * @example
 * ```html
 * <kendo-textbox>
 *   <ng-template kendoSuffixTemplate>
 *     <button kendoButton look="clear" icon="image"></button>
 *   </ng-template>
 * </kendo-textbox>
 * <kendo-multiselect [data]="data" [(ngModel)]="value">
 *   <ng-template kendoSuffixTemplate>
 *     <button kendoButton look="clear" icon="image"></button>
 *   </ng-template>
 * </kendo-multiselect>
 * ```
 */
export class SuffixTemplateDirective {
    templateRef;
    /**
     * Sets the `showSeparator` attribute of the `suffixTemplate`.
     *
     * @default false
     */
    set showSeparator(value) {
        this._showSeparator = value;
    }
    get showSeparator() {
        return this._showSeparator;
    }
    _showSeparator = false;
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SuffixTemplateDirective, deps: [{ token: i0.TemplateRef, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: SuffixTemplateDirective, isStandalone: true, selector: "[kendoSuffixTemplate]", inputs: { showSeparator: "showSeparator" }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SuffixTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[kendoSuffixTemplate]',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i0.TemplateRef, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { showSeparator: [{
                type: Input
            }] } });
