// @progress/kendo-angular-charts/esm2022/sankey/labels.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the configuration options of the Sankey labels
 * ([see example](slug:customization_sankey#toc-customizing-the-sankey-diagram-elements)).
 */
export class SankeyLabelsComponent extends SettingsComponent {
    configurationService;
    visible;
    font;
    color;
    align;
    position;
    padding;
    margin;
    border;
    offset;
    stroke;
    constructor(configurationService) {
        super('labels', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyLabelsComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SankeyLabelsComponent, isStandalone: true, selector: "kendo-sankey-labels", inputs: { visible: "visible", font: "font", color: "color", align: "align", position: "position", padding: "padding", margin: "margin", border: "border", offset: "offset", stroke: "stroke" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyLabelsComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-sankey-labels',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { visible: [{
                type: Input
            }], font: [{
                type: Input
            }], color: [{
                type: Input
            }], align: [{
                type: Input
            }], position: [{
                type: Input
            }], padding: [{
                type: Input
            }], margin: [{
                type: Input
            }], border: [{
                type: Input
            }], offset: [{
                type: Input
            }], stroke: [{
                type: Input
            }] } });
