// @progress/kendo-angular-charts/esm2022/sankey/nodes.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the configuration options of the Sankey nodes
 * ([see example](slug:customization_sankey#toc-customizing-the-sankey-diagram-elements)).
 */
export class SankeyNodesComponent extends SettingsComponent {
    configurationService;
    /**
     * Sets the color type of the link.
     * Use `static` for static link color determined by the link's `color` option.
     * Use `source` for link color that matches the source node color.
     * Use `target` for link color that matches the target node color.
     *
     * @default 'static'
     */
    colorType;
    color;
    opacity;
    offset;
    padding;
    width;
    align;
    focusHighlight;
    constructor(configurationService) {
        super('nodes', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyNodesComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SankeyNodesComponent, isStandalone: true, selector: "kendo-sankey-nodes", inputs: { colorType: "colorType", color: "color", opacity: "opacity", offset: "offset", padding: "padding", width: "width", align: "align", focusHighlight: "focusHighlight" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyNodesComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-sankey-nodes',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { colorType: [{
                type: Input
            }], color: [{
                type: Input
            }], opacity: [{
                type: Input
            }], offset: [{
                type: Input
            }], padding: [{
                type: Input
            }], width: [{
                type: Input
            }], align: [{
                type: Input
            }], focusHighlight: [{
                type: Input
            }] } });
