// @progress/kendo-angular-charts/esm2022/sankey/title.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ConfigurationService } from '../common/configuration.service';
import { SettingsComponent } from '../common/settings.component';
import * as i0 from "@angular/core";
import * as i1 from "../common/configuration.service";
/**
 * Represents the configuration options of the Sankey title or text
 * ([see example]({% slug title_sankey %})).
 *
 * @example
 * ```html
 * <kendo-sankey [data]="data">
 *     <kendo-sankey-title text="Title Text" [padding]="{bottom: 20}"></kendo-sankey-title>
 *     ...
 * </kendo-sankey>
 * ```
 */
export class SankeyTitleComponent extends SettingsComponent {
    configurationService;
    /**
     * Sets the alignment of the title.
     * Use `center` for middle alignment, `left` for left alignment, or `right` for right alignment.
     *
     * @default 'center'
     */
    align;
    /**
     * Sets the background color of the title. Accepts a valid CSS color string, including HEX and RGB.
     *
     * @default 'white'
     */
    background;
    border;
    color;
    /**
     * Sets the font of the title.
     *
     * @default '16px sans-serif'
     */
    font;
    margin;
    /**
     * Sets the padding of the title. A numeric value sets all paddings.
     *
     * @default 5
     */
    padding;
    /**
     * Sets the position of the title.
     * Use `bottom` to place the title at the bottom or `top` to place it at the top.
     *
     * @default 'top'
     */
    position;
    text;
    description;
    /**
     * Determines if the Sankey displays the title.
     *
     * @default true
     */
    visible;
    constructor(configurationService) {
        super('title', configurationService);
        this.configurationService = configurationService;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyTitleComponent, deps: [{ token: i1.ConfigurationService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "16.2.12", type: SankeyTitleComponent, isStandalone: true, selector: "kendo-sankey-title", inputs: { align: "align", background: "background", border: "border", color: "color", font: "font", margin: "margin", padding: "padding", position: "position", text: "text", description: "description", visible: "visible" }, usesInheritance: true, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: SankeyTitleComponent, decorators: [{
            type: Component,
            args: [{
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    selector: 'kendo-sankey-title',
                    template: '',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.ConfigurationService }]; }, propDecorators: { align: [{
                type: Input
            }], background: [{
                type: Input
            }], border: [{
                type: Input
            }], color: [{
                type: Input
            }], font: [{
                type: Input
            }], margin: [{
                type: Input
            }], padding: [{
                type: Input
            }], position: [{
                type: Input
            }], text: [{
                type: Input
            }], description: [{
                type: Input
            }], visible: [{
                type: Input
            }] } });
