// @progress/kendo-angular-dialog/esm2022/localization/custom-messages.component.mjs
/**-----------------------------------------------------------------------------------------
* Copyright © 2025 Progress Software Corporation. All rights reserved.
* Licensed under commercial license. See LICENSE.md in the project root for more information
*-------------------------------------------------------------------------------------------*/
import { Directive, forwardRef } from '@angular/core';
import { LocalizationService } from '@progress/kendo-angular-l10n';
import { Messages } from './messages';
import * as i0 from "@angular/core";
import * as i1 from "@progress/kendo-angular-l10n";
/**
 * Represents a component for customizing the messages in Dialog and Window components.
 *
 * Use the `CustomMessagesComponent` to override default messages with custom text. ([See example.](slug:globalization_dialogs#toc-internationalization)).
 *
 * @example
 * ```typescript
 * import { CustomMessagesComponent } from './custom-messages.component';
 *
 * @Component({
 *   selector: 'my-dialog',
 *   template: `
 *     <kendo-dialog>
 *       <kendo-dialog-messages
 *         [closeTitle]="'Close dialog'"
 *         [restoreTitle]="'Restore dialog'"
 *         [maximizeTitle]="'Maximize dialog'"
 *         [minimizeTitle]="'Minimize dialog'">
 *       </kendo-dialog-messages>
 *     </kendo-dialog>
 *   `
 * })
 * export class MyDialogComponent {}
 * ```
 */
export class CustomMessagesComponent extends Messages {
    service;
    constructor(service) {
        super();
        this.service = service;
    }
    get override() {
        return true;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CustomMessagesComponent, deps: [{ token: i1.LocalizationService }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.2.12", type: CustomMessagesComponent, isStandalone: true, selector: "kendo-dialog-messages, kendo-window-messages", providers: [
            {
                provide: Messages,
                useExisting: forwardRef(() => CustomMessagesComponent)
            }
        ], usesInheritance: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.2.12", ngImport: i0, type: CustomMessagesComponent, decorators: [{
            type: Directive,
            args: [{
                    providers: [
                        {
                            provide: Messages,
                            useExisting: forwardRef(() => CustomMessagesComponent)
                        }
                    ],
                    // eslint-disable-next-line @angular-eslint/directive-selector
                    selector: 'kendo-dialog-messages, kendo-window-messages',
                    standalone: true
                }]
        }], ctorParameters: function () { return [{ type: i1.LocalizationService }]; } });
